# Quantifying within-host evolution of H5N1 influenza in humans and poultry in Cambodia  

#### Louise H. Moncla<sup>1</sup>, Trevor Bedford<sup>1,2</sup>, Philippe Dussart<sup>3</sup>,  Srey Viseth Horm<sup>3</sup>, Sareth Rith<sup>3</sup>, Philippe Buchy<sup>4</sup>, Erik A Karlsson<sup>3</sup>,  Lifeng Li<sup>5,6</sup>, Yongmei Liu<sup>5,6</sup>, Huachen Zhu<sup>5,6</sup>, Yi Guan<sup>5,6</sup>, Thomas C. Friedrich<sup>7,8</sup>, Paul F. Horwood<sup>9,10</sup>

<sup>1</sup>Fred Hutchinson Cancer Research Center, Seattle, Washington, United States, <sup>2</sup>University of Washington, Seattle, Washington, United States, <sup>3</sup>Virology Unit, Institut Pasteur du Cambodge, Phnom Penh, Cambodia, <sup>4</sup>GlaxoSmithKline, Vaccines R&D, Singapore, Singapore,<sup>5</sup>Joint Influenza Research Centre (SUMC/HKU), Shantou University Medical College, Shantou, People's Republic of China,<sup>6</sup>State Key Laboratory of Emerging Infectious Diseases/Centre of Influenza Research, School of Public Health, The University of Hong Kong, Hong Kong, SAR, People's Republic of China,<sup>7</sup>Department of Pathobiological Sciences, University of Wisconsin School of Veterinary Medicine, Madison, WI, United States,<sup>8</sup>Wisconsin National Primate Research Center, Madison, WI, United States,<sup>9</sup>Papua New Guinea Institute of Medical Research, Goroka, Paula New Guinea,<sup>10</sup>Australian Institute of Tropical Health and Medicine, James Cook University, Cairns, Australia.

## Abstract

Avian influenza viruses (AIVs) periodically cross species barriers and infect humans. The likelihood that an AIV will evolve mammalian transmissibility depends on acquiring and selecting mutations during spillover. We analyze deep sequencing data from infected humans and ducks in Cambodia to examine H5N1 evolution during spillover. Viral populations in both species are predominated by low-frequency (<10%) variation shaped by purifying selection and genetic drift. Viruses from humans contain some human-adapting mutations (PB2 E627K, HA A150V, and HA Q238L), but these mutations remain low-frequency. Within-host variants are not enriched along phylogenetic branches leading to human infections. Our data show that H5N1 viruses generate putative human-adapting mutations during natural spillover infection. However, short infections, randomness, and purifying selection limit the evolutionary capacity of H5N1 viruses within-host. Applying evolutionary methods to sequence data, we reveal a detailed view of H5N1 adaptive potential, and develop a foundation for studying host-adaptation in other zoonotic viruses.

## Install

* Requires `baltic/baltic.py` to be available as a sister directory to `h5n1-cambodia`. BALTIC is available from [this GitHub repo](https://github.com/evogytis/baltic).

## Project structure

* [`auspice/`](auspice/): contains JSON trees viewable via Nextstrain
* [`data/`](data/): contains consensus genomes, within-host SNV calls and phylogenies
* [`figures/`](figures/): contains Jupyter notebooks to generate manuscript figures
* [`scripts`](scripts/): contains processing scripts
