{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 1c: Nonsynonymous and synonymous SNPs/site\n",
    "\n",
    "June 6, 2019 \n",
    "\n",
    "This notebook contains code for producing the diversity analyses as shown in **Supplementary Table 1** and for plotting **Figure 1c**. \n",
    "\n",
    "In both **Supplementary Table 1** and **Figure 1c**, the the mean nonsynonymous diversity (πN) and synonymous diversity (πS), per gene, per species is displayed. These were calculated for each gene in each individual sample by taking the number of synonymous or nonsynonymous SNPs called in that gene and sample, divided by the total number of synonymous or nonsynonymous sites present in that gene and sample. SNPs were called with Varscan using the [pipeline here](https://github.com/lmoncla/illumina_pipeline), requiring a minimum coverage of 100x, a minimum frequency of 1%, and a minimum qscore of Q30, and presence on forward and reverse reads. This is also described in the Methods of the manuscript. To define the number of synonymous and nonsynonymous sites per gene, I used [SNPGenie](https://github.com/chasewnelson/SNPGenie), which uses the [Nei-Gojobori (1986) method](https://academic.oup.com/mbe/article/3/5/418/988012). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, subprocess, glob, os, shutil, re, importlib, Bio\n",
    "from subprocess import call\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import rpy2\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "\n",
    "%load_ext rpy2.ipython "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define colors \n",
    "human_color = \"#C75643\"\n",
    "duck_color = \"#545AB7\"\n",
    "\n",
    "duck_nonsyn_color = \"#545AB7\"\n",
    "duck_syn_color = \"#98B4DA\"\n",
    "human_nonsyn_color = \"#C75643\"\n",
    "human_syn_color = \"#E6B692\"\n",
    "\n",
    "pd.set_option('display.max_rows', 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant calls file to load in\n",
    "variant_calls = \"/Users/lmoncla/src/h5n1-cambodia/data/within-host-variants-1%.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleid</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>reference_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>frequency(%)</th>\n",
       "      <th>frequency</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_H5</td>\n",
       "      <td>A/duck/Cambodia/381W11M4/2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>793</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala265Thr</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>3.28%</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_NP</td>\n",
       "      <td>A/duck/Cambodia/381W11M4/2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>384</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Gln117Arg</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>20.43%</td>\n",
       "      <td>0.2043</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A/duck/Cambodia/381W11M4/2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>939</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ala307Ala</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.55%</td>\n",
       "      <td>0.0455</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A/duck/Cambodia/381W11M4/2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Arg367Lys</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>19%</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A/duck/Cambodia/381W11M4/2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1608</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pro530Pro</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.38%</td>\n",
       "      <td>0.0438</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         sampleid  \\\n",
       "0  AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_H5   \n",
       "1  AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_NP   \n",
       "2  AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA   \n",
       "3  AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA   \n",
       "4  AJJ9KL706F510_A_duck_Cambodia_381W11M4_2013_PA   \n",
       "\n",
       "                          sample gene  reference_position reference_allele  \\\n",
       "0  A/duck/Cambodia/381W11M4/2013   HA                 793                G   \n",
       "1  A/duck/Cambodia/381W11M4/2013   NP                 384                A   \n",
       "2  A/duck/Cambodia/381W11M4/2013   PA                 939                A   \n",
       "3  A/duck/Cambodia/381W11M4/2013   PA                1118                G   \n",
       "4  A/duck/Cambodia/381W11M4/2013   PA                1608                G   \n",
       "\n",
       "  variant_allele coding_region_change synonymous_nonsynonymous frequency(%)  \\\n",
       "0              A            Ala265Thr            nonsynonymous        3.28%   \n",
       "1              G            Gln117Arg            nonsynonymous       20.43%   \n",
       "2              G            Ala307Ala               synonymous        4.55%   \n",
       "3              A            Arg367Lys            nonsynonymous          19%   \n",
       "4              A            Pro530Pro               synonymous        4.38%   \n",
       "\n",
       "   frequency  Unnamed: 10  \n",
       "0     0.0328          NaN  \n",
       "1     0.2043          NaN  \n",
       "2     0.0455          NaN  \n",
       "3     0.1900          NaN  \n",
       "4     0.0438          NaN  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_df = pd.read_csv(variant_calls, sep='\\t', header='infer')\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleid</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>reference_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>frequency(%)</th>\n",
       "      <th>frequency</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_H5</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>793</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala265Thr</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>3.28%</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_NP</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>384</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Gln117Arg</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>20.43%</td>\n",
       "      <td>0.2043</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>939</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ala307Ala</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.55%</td>\n",
       "      <td>0.0455</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Arg367Lys</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>19%</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1608</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pro530Pro</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.38%</td>\n",
       "      <td>0.0438</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sampleid                         sample gene  \\\n",
       "0  A_duck_Cambodia_381W11M4_2013_H5  A_duck_Cambodia_381W11M4_2013   HA   \n",
       "1  A_duck_Cambodia_381W11M4_2013_NP  A_duck_Cambodia_381W11M4_2013   NP   \n",
       "2  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "3  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "4  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "\n",
       "   reference_position reference_allele variant_allele coding_region_change  \\\n",
       "0                 793                G              A            Ala265Thr   \n",
       "1                 384                A              G            Gln117Arg   \n",
       "2                 939                A              G            Ala307Ala   \n",
       "3                1118                G              A            Arg367Lys   \n",
       "4                1608                G              A            Pro530Pro   \n",
       "\n",
       "  synonymous_nonsynonymous frequency(%)  frequency  Unnamed: 10  \n",
       "0            nonsynonymous        3.28%     0.0328          NaN  \n",
       "1            nonsynonymous       20.43%     0.2043          NaN  \n",
       "2               synonymous        4.55%     0.0455          NaN  \n",
       "3            nonsynonymous          19%     0.1900          NaN  \n",
       "4               synonymous        4.38%     0.0438          NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reformat the sampleid \n",
    "snps_df['sampleid'] = snps_df['sampleid'].str[14:]\n",
    "snps_df['sample'] = snps_df['sample'].str.replace(\"/\",\"_\")\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleid</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>reference_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>coding_region_change</th>\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>frequency(%)</th>\n",
       "      <th>frequency</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_H5</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>793</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala265Thr</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>3.28%</td>\n",
       "      <td>0.0328</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_NP</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>384</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Gln117Arg</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>20.43%</td>\n",
       "      <td>0.2043</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>939</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ala307Ala</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.55%</td>\n",
       "      <td>0.0455</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Arg367Lys</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>19%</td>\n",
       "      <td>0.1900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013_PA</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1608</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Pro530Pro</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>4.38%</td>\n",
       "      <td>0.0438</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sampleid                         sample gene  \\\n",
       "0  A_duck_Cambodia_381W11M4_2013_H5  A_duck_Cambodia_381W11M4_2013   HA   \n",
       "1  A_duck_Cambodia_381W11M4_2013_NP  A_duck_Cambodia_381W11M4_2013   NP   \n",
       "2  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "3  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "4  A_duck_Cambodia_381W11M4_2013_PA  A_duck_Cambodia_381W11M4_2013   PA   \n",
       "\n",
       "   reference_position reference_allele variant_allele coding_region_change  \\\n",
       "0                 793                G              A            Ala265Thr   \n",
       "1                 384                A              G            Gln117Arg   \n",
       "2                 939                A              G            Ala307Ala   \n",
       "3                1118                G              A            Arg367Lys   \n",
       "4                1608                G              A            Pro530Pro   \n",
       "\n",
       "  synonymous_nonsynonymous frequency(%)  frequency  Unnamed: 10  \n",
       "0            nonsynonymous        3.28%     0.0328          NaN  \n",
       "1            nonsynonymous       20.43%     0.2043          NaN  \n",
       "2               synonymous        4.55%     0.0455          NaN  \n",
       "3            nonsynonymous          19%     0.1900          NaN  \n",
       "4               synonymous        4.38%     0.0438          NaN  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get rid of the incorrect variant call due to a mismatched reference base\n",
    "snps_df = snps_df[snps_df['coding_region_change'] != 'Xaa240Gly']\n",
    "snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change NAs to neuramindase\n",
    "snps_df['gene'].fillna('neuraminidase', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>number_of_SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>nonsynonymous</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample gene synonymous_nonsynonymous  number_of_SNPs\n",
       "0  A_CAMBODIA_V0401301_2011   HA            nonsynonymous              12\n",
       "1  A_CAMBODIA_V0401301_2011   HA               synonymous               1\n",
       "2  A_CAMBODIA_V0401301_2011   M1            nonsynonymous               1\n",
       "4  A_CAMBODIA_V0401301_2011   NP            nonsynonymous               4\n",
       "5  A_CAMBODIA_V0401301_2011   NP               synonymous               2"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of SNPs per gene per sample and output as a dataframe\n",
    "counts = pd.DataFrame(snps_df.groupby([\"sample\", \"gene\",\"synonymous_nonsynonymous\"]).size())\n",
    "counts.reset_index(inplace=True)\n",
    "counts.columns = ['sample','gene','synonymous_nonsynonymous','number_of_SNPs']\n",
    "counts = counts[counts['synonymous_nonsynonymous'] != 'stop_gained']   # remove stop gained\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "synonymous_nonsynonymous                    sample gene  nonsynonymous  \\\n",
       "0                         A_CAMBODIA_V0401301_2011   HA           12.0   \n",
       "1                         A_CAMBODIA_V0401301_2011   M1            1.0   \n",
       "2                         A_CAMBODIA_V0401301_2011   NP            4.0   \n",
       "3                         A_CAMBODIA_V0401301_2011   PA            8.0   \n",
       "4                         A_CAMBODIA_V0401301_2011  PB1            4.0   \n",
       "\n",
       "synonymous_nonsynonymous  synonymous  \n",
       "0                                1.0  \n",
       "1                                0.0  \n",
       "2                                2.0  \n",
       "3                                1.0  \n",
       "4                                4.0  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# unmelt the synonymous nonsynonymous column\n",
    "counts = counts.pivot_table(index = ['sample','gene'], columns = 'synonymous_nonsynonymous', values='number_of_SNPs')\n",
    "counts.reset_index(inplace=True)\n",
    "counts = counts.fillna(0)\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "synonymous_nonsynonymous                    sample gene  nonsynonymous  \\\n",
       "0                         A_CAMBODIA_V0401301_2011   HA           12.0   \n",
       "1                         A_CAMBODIA_V0401301_2011   M1            1.0   \n",
       "2                         A_CAMBODIA_V0401301_2011   NP            4.0   \n",
       "3                         A_CAMBODIA_V0401301_2011   PA            8.0   \n",
       "4                         A_CAMBODIA_V0401301_2011  PB1            4.0   \n",
       "\n",
       "synonymous_nonsynonymous  synonymous species  \n",
       "0                                1.0   human  \n",
       "1                                0.0   human  \n",
       "2                                2.0   human  \n",
       "3                                1.0   human  \n",
       "4                                4.0   human  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add in a species column\n",
    "counts['species'] = counts['sample'].str.contains(\"duck\")\n",
    "counts['species'] = counts['species'].replace(True,\"duck\")\n",
    "counts['species'] = counts['species'].replace(False,\"human\")\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "74"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>synonymous_nonsynonymous</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>M2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PB2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>NP</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PB2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>PA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>NS1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>PA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>HA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>NP</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>duck</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "synonymous_nonsynonymous                         sample           gene  \\\n",
       "0                              A_CAMBODIA_V0401301_2011             HA   \n",
       "1                              A_CAMBODIA_V0401301_2011             M1   \n",
       "2                              A_CAMBODIA_V0401301_2011             NP   \n",
       "3                              A_CAMBODIA_V0401301_2011             PA   \n",
       "4                              A_CAMBODIA_V0401301_2011            PB1   \n",
       "5                              A_CAMBODIA_V0401301_2011            PB2   \n",
       "6                              A_CAMBODIA_V0401301_2011  neuraminidase   \n",
       "7                              A_CAMBODIA_V0417301_2011             HA   \n",
       "8                              A_CAMBODIA_V0417301_2011             M2   \n",
       "9                              A_CAMBODIA_V0417301_2011             NP   \n",
       "10                             A_CAMBODIA_V0417301_2011             PA   \n",
       "11                             A_CAMBODIA_V0417301_2011            PB2   \n",
       "12                             A_CAMBODIA_V0417301_2011  neuraminidase   \n",
       "13                             A_Cambodia_W0112303_2012             HA   \n",
       "14                             A_Cambodia_W0112303_2012             NP   \n",
       "15                             A_Cambodia_W0112303_2012             PA   \n",
       "16                             A_Cambodia_W0112303_2012            PB1   \n",
       "17                             A_Cambodia_W0112303_2012            PB2   \n",
       "18                             A_Cambodia_X0125302_2013             HA   \n",
       "19                             A_Cambodia_X0125302_2013             M1   \n",
       "20                             A_Cambodia_X0125302_2013             M2   \n",
       "21                             A_Cambodia_X0125302_2013            NS1   \n",
       "22                             A_Cambodia_X0125302_2013             PA   \n",
       "23                             A_Cambodia_X0125302_2013            PB1   \n",
       "24                             A_Cambodia_X0125302_2013            PB2   \n",
       "25                             A_Cambodia_X0125302_2013  neuraminidase   \n",
       "26                             A_Cambodia_X0128304_2013             HA   \n",
       "27                             A_Cambodia_X0128304_2013             M1   \n",
       "28                             A_Cambodia_X0128304_2013             M2   \n",
       "29                             A_Cambodia_X0128304_2013            NEP   \n",
       "30                             A_Cambodia_X0128304_2013             NP   \n",
       "31                             A_Cambodia_X0128304_2013            NS1   \n",
       "32                             A_Cambodia_X0128304_2013             PA   \n",
       "33                             A_Cambodia_X0128304_2013            PB2   \n",
       "34                             A_Cambodia_X0128304_2013  neuraminidase   \n",
       "35                             A_Cambodia_X0207301_2013             HA   \n",
       "36                             A_Cambodia_X0207301_2013             M1   \n",
       "37                             A_Cambodia_X0207301_2013            NEP   \n",
       "38                             A_Cambodia_X0207301_2013             NP   \n",
       "39                             A_Cambodia_X0207301_2013            NS1   \n",
       "40                             A_Cambodia_X0207301_2013             PA   \n",
       "41                             A_Cambodia_X0207301_2013            PB1   \n",
       "42                             A_Cambodia_X0207301_2013  neuraminidase   \n",
       "43                             A_Cambodia_X0219301_2013             NP   \n",
       "44                             A_Cambodia_X0219301_2013             PA   \n",
       "45                             A_Cambodia_X0219301_2013            PB2   \n",
       "46                             A_Cambodia_X1030304_2013             HA   \n",
       "47                             A_Cambodia_X1030304_2013             M1   \n",
       "48                             A_Cambodia_X1030304_2013             M2   \n",
       "49                             A_Cambodia_X1030304_2013            NS1   \n",
       "50                             A_Cambodia_X1030304_2013             PA   \n",
       "51                             A_Cambodia_X1030304_2013            PB1   \n",
       "52                             A_Cambodia_X1030304_2013            PB2   \n",
       "53                           A_duck_Cambodia_083D1_2011             HA   \n",
       "54                           A_duck_Cambodia_083D1_2011             M1   \n",
       "55                           A_duck_Cambodia_083D1_2011             NP   \n",
       "56                           A_duck_Cambodia_083D1_2011            PB1   \n",
       "57                           A_duck_Cambodia_083D1_2011  neuraminidase   \n",
       "58                        A_duck_Cambodia_381W11M4_2013             HA   \n",
       "59                        A_duck_Cambodia_381W11M4_2013             NP   \n",
       "60                        A_duck_Cambodia_381W11M4_2013             PA   \n",
       "61                        A_duck_Cambodia_381W11M4_2013            PB1   \n",
       "62                        A_duck_Cambodia_381W11M4_2013            PB2   \n",
       "63                         A_duck_Cambodia_PV027D1_2010             M1   \n",
       "64                         A_duck_Cambodia_PV027D1_2010             PA   \n",
       "65                         A_duck_Cambodia_PV027D1_2010            PB1   \n",
       "66                        A_duck_Cambodia_Y0224301_2014             HA   \n",
       "67                        A_duck_Cambodia_Y0224301_2014            NEP   \n",
       "68                        A_duck_Cambodia_Y0224301_2014            NS1   \n",
       "69                        A_duck_Cambodia_Y0224301_2014             PA   \n",
       "70                        A_duck_Cambodia_Y0224301_2014  neuraminidase   \n",
       "71                        A_duck_Cambodia_Y0224304_2014             HA   \n",
       "72                        A_duck_Cambodia_Y0224304_2014             NP   \n",
       "73                        A_duck_Cambodia_Y0224304_2014            PB2   \n",
       "\n",
       "synonymous_nonsynonymous  nonsynonymous  synonymous species  \n",
       "0                                  12.0         1.0   human  \n",
       "1                                   1.0         0.0   human  \n",
       "2                                   4.0         2.0   human  \n",
       "3                                   8.0         1.0   human  \n",
       "4                                   4.0         4.0   human  \n",
       "5                                   2.0         4.0   human  \n",
       "6                                   5.0         5.0   human  \n",
       "7                                   4.0         2.0   human  \n",
       "8                                   1.0         0.0   human  \n",
       "9                                   0.0         2.0   human  \n",
       "10                                  2.0         1.0   human  \n",
       "11                                  1.0         1.0   human  \n",
       "12                                  1.0         2.0   human  \n",
       "13                                  0.0         1.0   human  \n",
       "14                                  2.0         1.0   human  \n",
       "15                                  2.0         4.0   human  \n",
       "16                                  0.0         2.0   human  \n",
       "17                                  3.0         0.0   human  \n",
       "18                                  2.0         0.0   human  \n",
       "19                                  1.0         0.0   human  \n",
       "20                                  1.0         0.0   human  \n",
       "21                                  0.0         1.0   human  \n",
       "22                                  0.0         2.0   human  \n",
       "23                                  3.0         5.0   human  \n",
       "24                                  3.0         0.0   human  \n",
       "25                                  0.0         2.0   human  \n",
       "26                                  8.0         3.0   human  \n",
       "27                                  5.0         2.0   human  \n",
       "28                                  1.0         0.0   human  \n",
       "29                                  1.0         0.0   human  \n",
       "30                                  3.0         5.0   human  \n",
       "31                                  1.0         3.0   human  \n",
       "32                                  5.0         2.0   human  \n",
       "33                                  7.0         4.0   human  \n",
       "34                                  2.0         2.0   human  \n",
       "35                                  1.0         2.0   human  \n",
       "36                                  0.0         3.0   human  \n",
       "37                                  1.0         0.0   human  \n",
       "38                                  0.0         1.0   human  \n",
       "39                                  0.0         1.0   human  \n",
       "40                                  2.0         6.0   human  \n",
       "41                                  1.0         3.0   human  \n",
       "42                                  1.0         1.0   human  \n",
       "43                                  1.0         1.0   human  \n",
       "44                                  2.0         2.0   human  \n",
       "45                                  0.0         1.0   human  \n",
       "46                                  1.0         1.0   human  \n",
       "47                                  1.0         2.0   human  \n",
       "48                                  2.0         0.0   human  \n",
       "49                                  0.0         1.0   human  \n",
       "50                                  0.0         1.0   human  \n",
       "51                                  3.0         0.0   human  \n",
       "52                                  0.0         1.0   human  \n",
       "53                                  1.0         0.0    duck  \n",
       "54                                  1.0         0.0    duck  \n",
       "55                                  1.0         1.0    duck  \n",
       "56                                  0.0         1.0    duck  \n",
       "57                                  1.0         1.0    duck  \n",
       "58                                  1.0         0.0    duck  \n",
       "59                                  1.0         0.0    duck  \n",
       "60                                  1.0         2.0    duck  \n",
       "61                                  1.0         0.0    duck  \n",
       "62                                  0.0         4.0    duck  \n",
       "63                                  1.0         0.0    duck  \n",
       "64                                  0.0         1.0    duck  \n",
       "65                                  0.0         2.0    duck  \n",
       "66                                  0.0         1.0    duck  \n",
       "67                                  1.0         1.0    duck  \n",
       "68                                  2.0         0.0    duck  \n",
       "69                                  0.0         1.0    duck  \n",
       "70                                  0.0         1.0    duck  \n",
       "71                                  1.0         1.0    duck  \n",
       "72                                  3.0         5.0    duck  \n",
       "73                                  0.0         1.0    duck  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: {'gene': 'HA',\n",
       "  'nonsynonymous': 12.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 1: {'gene': 'M1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 2: {'gene': 'NP',\n",
       "  'nonsynonymous': 4.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 3: {'gene': 'PA',\n",
       "  'nonsynonymous': 8.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 4: {'gene': 'PB1',\n",
       "  'nonsynonymous': 4.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 4.0},\n",
       " 5: {'gene': 'PB2',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 4.0},\n",
       " 6: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 5.0,\n",
       "  'sample': 'A_CAMBODIA_V0401301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 5.0},\n",
       " 7: {'gene': 'HA',\n",
       "  'nonsynonymous': 4.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 8: {'gene': 'M2',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 9: {'gene': 'NP',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 10: {'gene': 'PA',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 11: {'gene': 'PB2',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 12: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_CAMBODIA_V0417301_2011',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 13: {'gene': 'HA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_W0112303_2012',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 14: {'gene': 'NP',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_W0112303_2012',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 15: {'gene': 'PA',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_W0112303_2012',\n",
       "  'species': 'human',\n",
       "  'synonymous': 4.0},\n",
       " 16: {'gene': 'PB1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_W0112303_2012',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 17: {'gene': 'PB2',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_Cambodia_W0112303_2012',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 18: {'gene': 'HA',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 19: {'gene': 'M1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 20: {'gene': 'M2',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 21: {'gene': 'NS1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 22: {'gene': 'PA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 23: {'gene': 'PB1',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 5.0},\n",
       " 24: {'gene': 'PB2',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 25: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0125302_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 26: {'gene': 'HA',\n",
       "  'nonsynonymous': 8.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 3.0},\n",
       " 27: {'gene': 'M1',\n",
       "  'nonsynonymous': 5.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 28: {'gene': 'M2',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 29: {'gene': 'NEP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 30: {'gene': 'NP',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 5.0},\n",
       " 31: {'gene': 'NS1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 3.0},\n",
       " 32: {'gene': 'PA',\n",
       "  'nonsynonymous': 5.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 33: {'gene': 'PB2',\n",
       "  'nonsynonymous': 7.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 4.0},\n",
       " 34: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_X0128304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 35: {'gene': 'HA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 36: {'gene': 'M1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 3.0},\n",
       " 37: {'gene': 'NEP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 38: {'gene': 'NP',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 39: {'gene': 'NS1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 40: {'gene': 'PA',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 6.0},\n",
       " 41: {'gene': 'PB1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 3.0},\n",
       " 42: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0207301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 43: {'gene': 'NP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X0219301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 44: {'gene': 'PA',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_X0219301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 45: {'gene': 'PB2',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X0219301_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 46: {'gene': 'HA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 47: {'gene': 'M1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 2.0},\n",
       " 48: {'gene': 'M2',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 49: {'gene': 'NS1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 50: {'gene': 'PA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 51: {'gene': 'PB1',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 0.0},\n",
       " 52: {'gene': 'PB2',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_Cambodia_X1030304_2013',\n",
       "  'species': 'human',\n",
       "  'synonymous': 1.0},\n",
       " 53: {'gene': 'HA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_083D1_2011',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 54: {'gene': 'M1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_083D1_2011',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 55: {'gene': 'NP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_083D1_2011',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 56: {'gene': 'PB1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_083D1_2011',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 57: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_083D1_2011',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 58: {'gene': 'HA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_381W11M4_2013',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 59: {'gene': 'NP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_381W11M4_2013',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 60: {'gene': 'PA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_381W11M4_2013',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 2.0},\n",
       " 61: {'gene': 'PB1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_381W11M4_2013',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 62: {'gene': 'PB2',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_381W11M4_2013',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 4.0},\n",
       " 63: {'gene': 'M1',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_PV027D1_2010',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 64: {'gene': 'PA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_PV027D1_2010',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 65: {'gene': 'PB1',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_PV027D1_2010',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 2.0},\n",
       " 66: {'gene': 'HA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224301_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 67: {'gene': 'NEP',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224301_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 68: {'gene': 'NS1',\n",
       "  'nonsynonymous': 2.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224301_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 0.0},\n",
       " 69: {'gene': 'PA',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224301_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 70: {'gene': 'neuraminidase',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224301_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 71: {'gene': 'HA',\n",
       "  'nonsynonymous': 1.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224304_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0},\n",
       " 72: {'gene': 'NP',\n",
       "  'nonsynonymous': 3.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224304_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 5.0},\n",
       " 73: {'gene': 'PB2',\n",
       "  'nonsynonymous': 0.0,\n",
       "  'sample': 'A_duck_Cambodia_Y0224304_2014',\n",
       "  'species': 'duck',\n",
       "  'synonymous': 1.0}}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write out counts to a dictionary \n",
    "counts_dict = counts.to_dict('index')\n",
    "counts_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in # of synonymous and nonsynonymous sites per gene per sample \n",
    "\n",
    "These results are from running SNPGenie and calculating pi. I will not be using the pi estimates, but will be using the number of synonymous and nonsynonymous sites per gene estimates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in SNPgenie pi data for each sample and gene to get the number of synonymous and nonsynonymous sites \n",
    "snp_genie_results = \"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/SNPGenie_diversity_estimates/1_percent_SNPs-2018-05-10/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites_dict = {}\n",
    "\n",
    "for f in glob.glob(snp_genie_results + \"*/SNPGenie_Results/product_results.txt\"):\n",
    "    with open(f,\"r\") as infile: \n",
    "        sampleid = f.replace(snp_genie_results, \"\").split(\"/\")[0][14:]\n",
    "        gene = sampleid.split(\"_\")[-1]\n",
    "        sample = sampleid.replace(\"_\"+gene, \"\")\n",
    "        \n",
    "        # change sample id to match the SNPs df \n",
    "        sampleid = sampleid.replace(\"M1\",\"MP\")\n",
    "        sampleid = sampleid.replace(\"M2\",\"MP\")\n",
    "        sampleid = sampleid.replace(\"NS1\",\"NS\")\n",
    "        sampleid = sampleid.replace(\"NEP\",\"NS\")\n",
    "        \n",
    "        gene = gene.replace(\"H5\",\"HA\")\n",
    "        gene = gene.replace(\"N1\",\"neuraminidase\")\n",
    "                \n",
    "        if sample not in sites_dict:\n",
    "            sites_dict[sample] = {gene:{}}\n",
    "                \n",
    "        for line in infile: \n",
    "            if \"file\" not in line: \n",
    "                Nsites = line.split()[6]\n",
    "                Ssites = line.split()[7]\n",
    "                \n",
    "        sites_dict[sample][gene] = {\"Nsites\": Nsites, \"Ssites\":Ssites}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A_CAMBODIA_V0401301_2011': ['M2', 'NS1', 'NEP'], 'A_CAMBODIA_V0417301_2011': ['PB1', 'M1', 'NS1', 'NEP'], 'A_Cambodia_W0112303_2012': ['neuraminidase', 'M1', 'M2', 'NS1', 'NEP'], 'A_Cambodia_X0125302_2013': ['NP', 'NEP'], 'A_Cambodia_X0207301_2013': ['PB2', 'M2'], 'A_Cambodia_X0219301_2013': ['PB1', 'HA', 'neuraminidase', 'M1', 'M2', 'NS1', 'NEP'], 'A_Cambodia_X1030304_2013': ['NP', 'neuraminidase', 'NEP'], 'A_duck_Cambodia_083D1_2011': ['PB2', 'PA', 'M2', 'NS1', 'NEP'], 'A_duck_Cambodia_381W11M4_2013': ['neuraminidase', 'M1', 'M2', 'NS1', 'NEP'], 'A_duck_Cambodia_PV027D1_2010': ['PB2', 'HA', 'NP', 'neuraminidase', 'M2', 'NS1', 'NEP'], 'A_duck_Cambodia_Y0224301_2014': ['PB2', 'PB1', 'NP', 'M1', 'M2'], 'A_duck_Cambodia_Y0224304_2014': ['PB1', 'PA', 'neuraminidase', 'M1', 'M2', 'NS1', 'NEP']}\n"
     ]
    }
   ],
   "source": [
    "# go through dictionary and add in missing rows for missing genes that do not have SNPs; still need to count these as 0 \n",
    "# the exception is PB1 for A/Cambodia/X0128304/2013 which just doesn't have data \n",
    "\n",
    "genes = ['PB2','PB1','PA','HA','NP','neuraminidase','M1','M2','NS1','NEP']\n",
    "has_snps = {}\n",
    "no_snps = {}\n",
    "\n",
    "for i in counts_dict:\n",
    "    sample = counts_dict[i]['sample']\n",
    "    gene = counts_dict[i]['gene']\n",
    "    \n",
    "    if sample not in has_snps: \n",
    "        has_snps[sample] = []\n",
    "    \n",
    "    has_snps[sample].append(gene)\n",
    "\n",
    "# now loop through to output ones without snps \n",
    "for s in has_snps: \n",
    "    for g in genes: \n",
    "        if g not in has_snps[s]:\n",
    "            if g == 'PB1' and s == 'A_Cambodia_X0128304_2013':\n",
    "                pass\n",
    "            else:\n",
    "                if s not in no_snps:\n",
    "                    no_snps[s] = []\n",
    "                    no_snps[s].append(g)\n",
    "                elif s in no_snps:\n",
    "                    no_snps[s].append(g)\n",
    "            \n",
    "print(no_snps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "      <th>N_per_site</th>\n",
       "      <th>S_per_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample gene nonsynonymous synonymous species Nsites  \\\n",
       "0  A_CAMBODIA_V0401301_2011   M2             0          0   human      0   \n",
       "0  A_CAMBODIA_V0401301_2011  NS1             0          0   human      0   \n",
       "0  A_CAMBODIA_V0401301_2011  NEP             0          0   human      0   \n",
       "0  A_CAMBODIA_V0417301_2011  PB1             0          0   human      0   \n",
       "0  A_CAMBODIA_V0417301_2011   M1             0          0   human      0   \n",
       "\n",
       "  Ssites N_per_site S_per_site  \n",
       "0      0          0          0  \n",
       "0      0          0          0  \n",
       "0      0          0          0  \n",
       "0      0          0          0  \n",
       "0      0          0          0  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write nosnps out to a dataframe \n",
    "no_snps_df = pd.DataFrame()\n",
    "\n",
    "for s in no_snps: \n",
    "    for g in no_snps[s]:\n",
    "        sampleid = s + \"_\" + g\n",
    "        if \"duck\" in s:\n",
    "            species = \"duck\"\n",
    "        else:\n",
    "            species = \"human\"\n",
    "                \n",
    "        d = pd.DataFrame(np.array([[s,g, 0,0,species,0,0,0,0]]))\n",
    "        d.columns = ['sample','gene','nonsynonymous','synonymous','species','Nsites','Ssites','N_per_site','S_per_site']\n",
    "\n",
    "        no_snps_df = no_snps_df.append(d)\n",
    "\n",
    "no_snps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1332.74039747032</td>\n",
       "      <td>370.650906877508</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>HA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>578.515593779322</td>\n",
       "      <td>177.484406220678</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>227</td>\n",
       "      <td>64</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1050.11942257218</td>\n",
       "      <td>296.880577427822</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>286.009963768116</td>\n",
       "      <td>76.990036231884</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NEP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Nsites            Ssites                    sample           gene\n",
       "0  1332.74039747032  370.650906877508  A_Cambodia_X0128304_2013             HA\n",
       "0  578.515593779322  177.484406220678  A_Cambodia_X0128304_2013             M1\n",
       "0               227                64  A_Cambodia_X0128304_2013             M2\n",
       "0  1050.11942257218  296.880577427822  A_Cambodia_X0128304_2013  neuraminidase\n",
       "0  286.009963768116   76.990036231884  A_Cambodia_X0128304_2013            NEP"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert the nonsynonymous and synonymous sites per gene to a dataframe\n",
    "sites_df = pd.DataFrame()\n",
    "\n",
    "for s in sites_dict: \n",
    "    \n",
    "    for g in sites_dict[s]:\n",
    "        df = pd.DataFrame.from_dict(sites_dict[s][g], orient=\"index\").T\n",
    "        df['sample'] = s\n",
    "        df['gene'] = g\n",
    "    \n",
    "        sites_df = sites_df.append(df)\n",
    "\n",
    "sites_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sites_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge together the SNP counts dataframe and the synonymous/nonsynonymous lengths dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1148.01647108678</td>\n",
       "      <td>345.983528913223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1684.40017392789</td>\n",
       "      <td>463.421018125091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1767.9268235821</td>\n",
       "      <td>503.073176417901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample gene  nonsynonymous  synonymous species  \\\n",
       "0  A_CAMBODIA_V0401301_2011   HA           12.0         1.0   human   \n",
       "1  A_CAMBODIA_V0401301_2011   M1            1.0         0.0   human   \n",
       "2  A_CAMBODIA_V0401301_2011   NP            4.0         2.0   human   \n",
       "3  A_CAMBODIA_V0401301_2011   PA            8.0         1.0   human   \n",
       "4  A_CAMBODIA_V0401301_2011  PB1            4.0         4.0   human   \n",
       "\n",
       "             Nsites            Ssites  \n",
       "0   1332.9534942423  371.046505757698  \n",
       "1   1332.9534942423  371.046505757698  \n",
       "2  1148.01647108678  345.983528913223  \n",
       "3  1684.40017392789  463.421018125091  \n",
       "4   1767.9268235821  503.073176417901  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge the dataframes including the counts of SNPs and the synonymous and nonsynonymou sites\n",
    "merged = counts.merge(sites_df, on=['sample','gene'], how='outer')\n",
    "merged = merged.dropna()\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "      <th>N_per_site</th>\n",
       "      <th>S_per_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.009003</td>\n",
       "      <td>0.002695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1148.01647108678</td>\n",
       "      <td>345.983528913223</td>\n",
       "      <td>0.003484</td>\n",
       "      <td>0.005781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1684.40017392789</td>\n",
       "      <td>463.421018125091</td>\n",
       "      <td>0.004749</td>\n",
       "      <td>0.002158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>human</td>\n",
       "      <td>1767.9268235821</td>\n",
       "      <td>503.073176417901</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.007951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample gene  nonsynonymous  synonymous species  \\\n",
       "0  A_CAMBODIA_V0401301_2011   HA           12.0         1.0   human   \n",
       "1  A_CAMBODIA_V0401301_2011   M1            1.0         0.0   human   \n",
       "2  A_CAMBODIA_V0401301_2011   NP            4.0         2.0   human   \n",
       "3  A_CAMBODIA_V0401301_2011   PA            8.0         1.0   human   \n",
       "4  A_CAMBODIA_V0401301_2011  PB1            4.0         4.0   human   \n",
       "\n",
       "             Nsites            Ssites  N_per_site  S_per_site  \n",
       "0   1332.9534942423  371.046505757698    0.009003    0.002695  \n",
       "1   1332.9534942423  371.046505757698    0.000750    0.000000  \n",
       "2  1148.01647108678  345.983528913223    0.003484    0.005781  \n",
       "3  1684.40017392789  463.421018125091    0.004749    0.002158  \n",
       "4   1767.9268235821  503.073176417901    0.002263    0.007951  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add a column for nonsynonymous and synonymous per site \n",
    "merged['N_per_site'] = merged['nonsynonymous'].astype(float) / merged['Nsites'].astype(float)\n",
    "merged['S_per_site'] = merged['synonymous'].astype(float) / merged['Ssites'].astype(float)\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add in the no snps dataframe\n",
    "merged = merged.append(no_snps_df)\n",
    "#merged.dropna(inplace=True)\n",
    "merged.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "      <th>N_per_site</th>\n",
       "      <th>S_per_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.00900256</td>\n",
       "      <td>0.00269508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.000750214</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1148.01647108678</td>\n",
       "      <td>345.983528913223</td>\n",
       "      <td>0.00348427</td>\n",
       "      <td>0.00578062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1684.40017392789</td>\n",
       "      <td>463.421018125091</td>\n",
       "      <td>0.00474947</td>\n",
       "      <td>0.00215787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>human</td>\n",
       "      <td>1767.9268235821</td>\n",
       "      <td>503.073176417901</td>\n",
       "      <td>0.00226254</td>\n",
       "      <td>0.00795113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>human</td>\n",
       "      <td>1759.40919088767</td>\n",
       "      <td>517.528122545167</td>\n",
       "      <td>0.00113675</td>\n",
       "      <td>0.00772905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>human</td>\n",
       "      <td>1049.14877034739</td>\n",
       "      <td>297.851229652607</td>\n",
       "      <td>0.00476577</td>\n",
       "      <td>0.0167869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.78501496308</td>\n",
       "      <td>371.21498503692</td>\n",
       "      <td>0.00300123</td>\n",
       "      <td>0.00538771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>M2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>227.806260575296</td>\n",
       "      <td>63.1937394247039</td>\n",
       "      <td>0.00438969</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1147.62311557789</td>\n",
       "      <td>346.37688442211</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00577406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1683.83333333334</td>\n",
       "      <td>464.166666666666</td>\n",
       "      <td>0.00118777</td>\n",
       "      <td>0.0021544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PB2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1757.69589041096</td>\n",
       "      <td>519.304109589039</td>\n",
       "      <td>0.000568927</td>\n",
       "      <td>0.00192565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1049.51340033501</td>\n",
       "      <td>297.486599664992</td>\n",
       "      <td>0.000952823</td>\n",
       "      <td>0.00672299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>HA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1333.66666666667</td>\n",
       "      <td>370.333333333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00270027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>NP</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1148.13882383287</td>\n",
       "      <td>345.861176167126</td>\n",
       "      <td>0.00174195</td>\n",
       "      <td>0.00289133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PA</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>human</td>\n",
       "      <td>1683.69121077319</td>\n",
       "      <td>464.308789226808</td>\n",
       "      <td>0.00118787</td>\n",
       "      <td>0.00861496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1768.06094307243</td>\n",
       "      <td>502.939056927572</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00397662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>PB2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1757.4705882353</td>\n",
       "      <td>519.529411764704</td>\n",
       "      <td>0.001707</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.17396091385</td>\n",
       "      <td>371.826039086154</td>\n",
       "      <td>0.00150131</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>578.666666666666</td>\n",
       "      <td>177.333333333333</td>\n",
       "      <td>0.00172811</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>227.018087855297</td>\n",
       "      <td>63.9819121447029</td>\n",
       "      <td>0.00440494</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>521.166666666667</td>\n",
       "      <td>153.833333333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00650054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1682</td>\n",
       "      <td>465.999999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00429185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>human</td>\n",
       "      <td>1766.50153834872</td>\n",
       "      <td>504.498461651281</td>\n",
       "      <td>0.00169827</td>\n",
       "      <td>0.00991083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1751.55580621173</td>\n",
       "      <td>525.444193788278</td>\n",
       "      <td>0.00171276</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1049.5</td>\n",
       "      <td>297.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00672269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.74039747032</td>\n",
       "      <td>370.650906877508</td>\n",
       "      <td>0.00600267</td>\n",
       "      <td>0.00809387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>578.515593779322</td>\n",
       "      <td>177.484406220678</td>\n",
       "      <td>0.00864281</td>\n",
       "      <td>0.0112686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>227</td>\n",
       "      <td>64</td>\n",
       "      <td>0.00440529</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>286.009963768116</td>\n",
       "      <td>76.990036231884</td>\n",
       "      <td>0.00349638</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>30</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>human</td>\n",
       "      <td>1145.6420481738</td>\n",
       "      <td>348.357951826201</td>\n",
       "      <td>0.00261862</td>\n",
       "      <td>0.0143531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>31</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>human</td>\n",
       "      <td>520.5</td>\n",
       "      <td>154.5</td>\n",
       "      <td>0.00192123</td>\n",
       "      <td>0.0194175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>32</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1682</td>\n",
       "      <td>465.999999999999</td>\n",
       "      <td>0.00297265</td>\n",
       "      <td>0.00429185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>33</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>human</td>\n",
       "      <td>1751.86491175721</td>\n",
       "      <td>525.135088242794</td>\n",
       "      <td>0.00399574</td>\n",
       "      <td>0.00761709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>34</td>\n",
       "      <td>A_Cambodia_X0128304_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1050.11942257218</td>\n",
       "      <td>296.880577427822</td>\n",
       "      <td>0.00190455</td>\n",
       "      <td>0.00673672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>35</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1333.21212121212</td>\n",
       "      <td>370.787878787878</td>\n",
       "      <td>0.000750068</td>\n",
       "      <td>0.00539392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>36</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>human</td>\n",
       "      <td>578.666666666666</td>\n",
       "      <td>177.333333333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0169173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>37</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>285.969782813975</td>\n",
       "      <td>77.0302171860245</td>\n",
       "      <td>0.00349687</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>38</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1145.6600177305</td>\n",
       "      <td>348.339982269503</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00287076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>39</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>520.507554296507</td>\n",
       "      <td>154.492445703494</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00647281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>40</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>human</td>\n",
       "      <td>1681.97389375589</td>\n",
       "      <td>466.026106244114</td>\n",
       "      <td>0.00118908</td>\n",
       "      <td>0.0128748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>41</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>human</td>\n",
       "      <td>1766.50000000001</td>\n",
       "      <td>504.499999999998</td>\n",
       "      <td>0.000566091</td>\n",
       "      <td>0.00594648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>42</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1047.5</td>\n",
       "      <td>296.5</td>\n",
       "      <td>0.000954654</td>\n",
       "      <td>0.00337268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>43</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1145.66666666667</td>\n",
       "      <td>348.333333333333</td>\n",
       "      <td>0.000872854</td>\n",
       "      <td>0.00287081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>44</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1682.98511904762</td>\n",
       "      <td>465.01488095238</td>\n",
       "      <td>0.00118836</td>\n",
       "      <td>0.00430094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>45</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1751.00000000001</td>\n",
       "      <td>525.999999999998</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00190114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>46</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.12529832936</td>\n",
       "      <td>371.874701670644</td>\n",
       "      <td>0.00075068</td>\n",
       "      <td>0.00268908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>47</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>578.648484848485</td>\n",
       "      <td>177.351515151515</td>\n",
       "      <td>0.00172816</td>\n",
       "      <td>0.011277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>48</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>226.990616621984</td>\n",
       "      <td>64.0093833780161</td>\n",
       "      <td>0.00881094</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>49</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>520.5</td>\n",
       "      <td>154.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00647249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>50</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1682.16666666667</td>\n",
       "      <td>465.833333333332</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00214669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>51</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1766.33993226257</td>\n",
       "      <td>504.660067737432</td>\n",
       "      <td>0.00169843</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>52</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1751.19242020513</td>\n",
       "      <td>525.807579794871</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00190184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>53</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>1332.33333333333</td>\n",
       "      <td>371.666666666666</td>\n",
       "      <td>0.000750563</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>54</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>577.112612612612</td>\n",
       "      <td>178.806306306306</td>\n",
       "      <td>0.00173276</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>55</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1148</td>\n",
       "      <td>345.999999999999</td>\n",
       "      <td>0.00087108</td>\n",
       "      <td>0.00289017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>56</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1766.16666666667</td>\n",
       "      <td>504.833333333332</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00198085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>57</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1049.5</td>\n",
       "      <td>297.5</td>\n",
       "      <td>0.000952835</td>\n",
       "      <td>0.00336134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>58</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>1330.33333333333</td>\n",
       "      <td>370.666666666666</td>\n",
       "      <td>0.000751691</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>59</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>1146.46241134752</td>\n",
       "      <td>347.537588652482</td>\n",
       "      <td>0.000872248</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>60</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PA</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "      <td>1681.73001508296</td>\n",
       "      <td>466.269984917043</td>\n",
       "      <td>0.000594626</td>\n",
       "      <td>0.00428936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>61</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>1766.1245210728</td>\n",
       "      <td>504.875478927201</td>\n",
       "      <td>0.000566211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>duck</td>\n",
       "      <td>1751.33333333334</td>\n",
       "      <td>525.666666666665</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00760938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>63</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>577.166666666667</td>\n",
       "      <td>178.833333333333</td>\n",
       "      <td>0.0017326</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>64</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1683.5</td>\n",
       "      <td>464.499999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00215285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>65</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>duck</td>\n",
       "      <td>1766.99449944995</td>\n",
       "      <td>504.005500550053</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00396821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>66</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>HA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1334</td>\n",
       "      <td>369.999999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0027027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>67</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>NEP</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>285.985185185185</td>\n",
       "      <td>77.0148148148148</td>\n",
       "      <td>0.00349668</td>\n",
       "      <td>0.0129845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>68</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>NS1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>520.5</td>\n",
       "      <td>154.5</td>\n",
       "      <td>0.00384246</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>69</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1682</td>\n",
       "      <td>465.999999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00214592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1049.66666666666</td>\n",
       "      <td>297.333333333334</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00336323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>71</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>HA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1328.85441370224</td>\n",
       "      <td>372.145586297759</td>\n",
       "      <td>0.000752528</td>\n",
       "      <td>0.00268712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>72</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>NP</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>duck</td>\n",
       "      <td>1145.70741842189</td>\n",
       "      <td>348.292581578106</td>\n",
       "      <td>0.00261847</td>\n",
       "      <td>0.0143557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>73</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>duck</td>\n",
       "      <td>1751.83333333334</td>\n",
       "      <td>525.166666666665</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00190416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0417301_2011</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_W0112303_2012</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0125302_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0207301_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>HA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X0219301_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>0</td>\n",
       "      <td>A_Cambodia_X1030304_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_083D1_2011</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_381W11M4_2013</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>HA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_PV027D1_2010</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>PB2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>NP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224301_2014</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>PA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>M2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>NS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>0</td>\n",
       "      <td>A_duck_Cambodia_Y0224304_2014</td>\n",
       "      <td>NEP</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>duck</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index                         sample           gene nonsynonymous  \\\n",
       "0        0       A_CAMBODIA_V0401301_2011             HA            12   \n",
       "1        1       A_CAMBODIA_V0401301_2011             M1             1   \n",
       "2        2       A_CAMBODIA_V0401301_2011             NP             4   \n",
       "3        3       A_CAMBODIA_V0401301_2011             PA             8   \n",
       "4        4       A_CAMBODIA_V0401301_2011            PB1             4   \n",
       "5        5       A_CAMBODIA_V0401301_2011            PB2             2   \n",
       "6        6       A_CAMBODIA_V0401301_2011  neuraminidase             5   \n",
       "7        7       A_CAMBODIA_V0417301_2011             HA             4   \n",
       "8        8       A_CAMBODIA_V0417301_2011             M2             1   \n",
       "9        9       A_CAMBODIA_V0417301_2011             NP             0   \n",
       "10      10       A_CAMBODIA_V0417301_2011             PA             2   \n",
       "11      11       A_CAMBODIA_V0417301_2011            PB2             1   \n",
       "12      12       A_CAMBODIA_V0417301_2011  neuraminidase             1   \n",
       "13      13       A_Cambodia_W0112303_2012             HA             0   \n",
       "14      14       A_Cambodia_W0112303_2012             NP             2   \n",
       "15      15       A_Cambodia_W0112303_2012             PA             2   \n",
       "16      16       A_Cambodia_W0112303_2012            PB1             0   \n",
       "17      17       A_Cambodia_W0112303_2012            PB2             3   \n",
       "18      18       A_Cambodia_X0125302_2013             HA             2   \n",
       "19      19       A_Cambodia_X0125302_2013             M1             1   \n",
       "20      20       A_Cambodia_X0125302_2013             M2             1   \n",
       "21      21       A_Cambodia_X0125302_2013            NS1             0   \n",
       "22      22       A_Cambodia_X0125302_2013             PA             0   \n",
       "23      23       A_Cambodia_X0125302_2013            PB1             3   \n",
       "24      24       A_Cambodia_X0125302_2013            PB2             3   \n",
       "25      25       A_Cambodia_X0125302_2013  neuraminidase             0   \n",
       "26      26       A_Cambodia_X0128304_2013             HA             8   \n",
       "27      27       A_Cambodia_X0128304_2013             M1             5   \n",
       "28      28       A_Cambodia_X0128304_2013             M2             1   \n",
       "29      29       A_Cambodia_X0128304_2013            NEP             1   \n",
       "30      30       A_Cambodia_X0128304_2013             NP             3   \n",
       "31      31       A_Cambodia_X0128304_2013            NS1             1   \n",
       "32      32       A_Cambodia_X0128304_2013             PA             5   \n",
       "33      33       A_Cambodia_X0128304_2013            PB2             7   \n",
       "34      34       A_Cambodia_X0128304_2013  neuraminidase             2   \n",
       "35      35       A_Cambodia_X0207301_2013             HA             1   \n",
       "36      36       A_Cambodia_X0207301_2013             M1             0   \n",
       "37      37       A_Cambodia_X0207301_2013            NEP             1   \n",
       "38      38       A_Cambodia_X0207301_2013             NP             0   \n",
       "39      39       A_Cambodia_X0207301_2013            NS1             0   \n",
       "40      40       A_Cambodia_X0207301_2013             PA             2   \n",
       "41      41       A_Cambodia_X0207301_2013            PB1             1   \n",
       "42      42       A_Cambodia_X0207301_2013  neuraminidase             1   \n",
       "43      43       A_Cambodia_X0219301_2013             NP             1   \n",
       "44      44       A_Cambodia_X0219301_2013             PA             2   \n",
       "45      45       A_Cambodia_X0219301_2013            PB2             0   \n",
       "46      46       A_Cambodia_X1030304_2013             HA             1   \n",
       "47      47       A_Cambodia_X1030304_2013             M1             1   \n",
       "48      48       A_Cambodia_X1030304_2013             M2             2   \n",
       "49      49       A_Cambodia_X1030304_2013            NS1             0   \n",
       "50      50       A_Cambodia_X1030304_2013             PA             0   \n",
       "51      51       A_Cambodia_X1030304_2013            PB1             3   \n",
       "52      52       A_Cambodia_X1030304_2013            PB2             0   \n",
       "53      53     A_duck_Cambodia_083D1_2011             HA             1   \n",
       "54      54     A_duck_Cambodia_083D1_2011             M1             1   \n",
       "55      55     A_duck_Cambodia_083D1_2011             NP             1   \n",
       "56      56     A_duck_Cambodia_083D1_2011            PB1             0   \n",
       "57      57     A_duck_Cambodia_083D1_2011  neuraminidase             1   \n",
       "58      58  A_duck_Cambodia_381W11M4_2013             HA             1   \n",
       "59      59  A_duck_Cambodia_381W11M4_2013             NP             1   \n",
       "60      60  A_duck_Cambodia_381W11M4_2013             PA             1   \n",
       "61      61  A_duck_Cambodia_381W11M4_2013            PB1             1   \n",
       "62      62  A_duck_Cambodia_381W11M4_2013            PB2             0   \n",
       "63      63   A_duck_Cambodia_PV027D1_2010             M1             1   \n",
       "64      64   A_duck_Cambodia_PV027D1_2010             PA             0   \n",
       "65      65   A_duck_Cambodia_PV027D1_2010            PB1             0   \n",
       "66      66  A_duck_Cambodia_Y0224301_2014             HA             0   \n",
       "67      67  A_duck_Cambodia_Y0224301_2014            NEP             1   \n",
       "68      68  A_duck_Cambodia_Y0224301_2014            NS1             2   \n",
       "69      69  A_duck_Cambodia_Y0224301_2014             PA             0   \n",
       "70      70  A_duck_Cambodia_Y0224301_2014  neuraminidase             0   \n",
       "71      71  A_duck_Cambodia_Y0224304_2014             HA             1   \n",
       "72      72  A_duck_Cambodia_Y0224304_2014             NP             3   \n",
       "73      73  A_duck_Cambodia_Y0224304_2014            PB2             0   \n",
       "74       0       A_CAMBODIA_V0401301_2011             M2             0   \n",
       "75       0       A_CAMBODIA_V0401301_2011            NS1             0   \n",
       "76       0       A_CAMBODIA_V0401301_2011            NEP             0   \n",
       "77       0       A_CAMBODIA_V0417301_2011            PB1             0   \n",
       "78       0       A_CAMBODIA_V0417301_2011             M1             0   \n",
       "79       0       A_CAMBODIA_V0417301_2011            NS1             0   \n",
       "80       0       A_CAMBODIA_V0417301_2011            NEP             0   \n",
       "81       0       A_Cambodia_W0112303_2012  neuraminidase             0   \n",
       "82       0       A_Cambodia_W0112303_2012             M1             0   \n",
       "83       0       A_Cambodia_W0112303_2012             M2             0   \n",
       "84       0       A_Cambodia_W0112303_2012            NS1             0   \n",
       "85       0       A_Cambodia_W0112303_2012            NEP             0   \n",
       "86       0       A_Cambodia_X0125302_2013             NP             0   \n",
       "87       0       A_Cambodia_X0125302_2013            NEP             0   \n",
       "88       0       A_Cambodia_X0207301_2013            PB2             0   \n",
       "89       0       A_Cambodia_X0207301_2013             M2             0   \n",
       "90       0       A_Cambodia_X0219301_2013            PB1             0   \n",
       "91       0       A_Cambodia_X0219301_2013             HA             0   \n",
       "92       0       A_Cambodia_X0219301_2013  neuraminidase             0   \n",
       "93       0       A_Cambodia_X0219301_2013             M1             0   \n",
       "94       0       A_Cambodia_X0219301_2013             M2             0   \n",
       "95       0       A_Cambodia_X0219301_2013            NS1             0   \n",
       "96       0       A_Cambodia_X0219301_2013            NEP             0   \n",
       "97       0       A_Cambodia_X1030304_2013             NP             0   \n",
       "98       0       A_Cambodia_X1030304_2013  neuraminidase             0   \n",
       "99       0       A_Cambodia_X1030304_2013            NEP             0   \n",
       "100      0     A_duck_Cambodia_083D1_2011            PB2             0   \n",
       "101      0     A_duck_Cambodia_083D1_2011             PA             0   \n",
       "102      0     A_duck_Cambodia_083D1_2011             M2             0   \n",
       "103      0     A_duck_Cambodia_083D1_2011            NS1             0   \n",
       "104      0     A_duck_Cambodia_083D1_2011            NEP             0   \n",
       "105      0  A_duck_Cambodia_381W11M4_2013  neuraminidase             0   \n",
       "106      0  A_duck_Cambodia_381W11M4_2013             M1             0   \n",
       "107      0  A_duck_Cambodia_381W11M4_2013             M2             0   \n",
       "108      0  A_duck_Cambodia_381W11M4_2013            NS1             0   \n",
       "109      0  A_duck_Cambodia_381W11M4_2013            NEP             0   \n",
       "110      0   A_duck_Cambodia_PV027D1_2010            PB2             0   \n",
       "111      0   A_duck_Cambodia_PV027D1_2010             HA             0   \n",
       "112      0   A_duck_Cambodia_PV027D1_2010             NP             0   \n",
       "113      0   A_duck_Cambodia_PV027D1_2010  neuraminidase             0   \n",
       "114      0   A_duck_Cambodia_PV027D1_2010             M2             0   \n",
       "115      0   A_duck_Cambodia_PV027D1_2010            NS1             0   \n",
       "116      0   A_duck_Cambodia_PV027D1_2010            NEP             0   \n",
       "117      0  A_duck_Cambodia_Y0224301_2014            PB2             0   \n",
       "118      0  A_duck_Cambodia_Y0224301_2014            PB1             0   \n",
       "119      0  A_duck_Cambodia_Y0224301_2014             NP             0   \n",
       "120      0  A_duck_Cambodia_Y0224301_2014             M1             0   \n",
       "121      0  A_duck_Cambodia_Y0224301_2014             M2             0   \n",
       "122      0  A_duck_Cambodia_Y0224304_2014            PB1             0   \n",
       "123      0  A_duck_Cambodia_Y0224304_2014             PA             0   \n",
       "124      0  A_duck_Cambodia_Y0224304_2014  neuraminidase             0   \n",
       "125      0  A_duck_Cambodia_Y0224304_2014             M1             0   \n",
       "126      0  A_duck_Cambodia_Y0224304_2014             M2             0   \n",
       "127      0  A_duck_Cambodia_Y0224304_2014            NS1             0   \n",
       "128      0  A_duck_Cambodia_Y0224304_2014            NEP             0   \n",
       "\n",
       "    synonymous species            Nsites            Ssites   N_per_site  \\\n",
       "0            1   human   1332.9534942423  371.046505757698   0.00900256   \n",
       "1            0   human   1332.9534942423  371.046505757698  0.000750214   \n",
       "2            2   human  1148.01647108678  345.983528913223   0.00348427   \n",
       "3            1   human  1684.40017392789  463.421018125091   0.00474947   \n",
       "4            4   human   1767.9268235821  503.073176417901   0.00226254   \n",
       "5            4   human  1759.40919088767  517.528122545167   0.00113675   \n",
       "6            5   human  1049.14877034739  297.851229652607   0.00476577   \n",
       "7            2   human  1332.78501496308   371.21498503692   0.00300123   \n",
       "8            0   human  227.806260575296  63.1937394247039   0.00438969   \n",
       "9            2   human  1147.62311557789   346.37688442211            0   \n",
       "10           1   human  1683.83333333334  464.166666666666   0.00118777   \n",
       "11           1   human  1757.69589041096  519.304109589039  0.000568927   \n",
       "12           2   human  1049.51340033501  297.486599664992  0.000952823   \n",
       "13           1   human  1333.66666666667  370.333333333333            0   \n",
       "14           1   human  1148.13882383287  345.861176167126   0.00174195   \n",
       "15           4   human  1683.69121077319  464.308789226808   0.00118787   \n",
       "16           2   human  1768.06094307243  502.939056927572            0   \n",
       "17           0   human   1757.4705882353  519.529411764704     0.001707   \n",
       "18           0   human  1332.17396091385  371.826039086154   0.00150131   \n",
       "19           0   human  578.666666666666  177.333333333333   0.00172811   \n",
       "20           0   human  227.018087855297  63.9819121447029   0.00440494   \n",
       "21           1   human  521.166666666667  153.833333333333            0   \n",
       "22           2   human              1682  465.999999999999            0   \n",
       "23           5   human  1766.50153834872  504.498461651281   0.00169827   \n",
       "24           0   human  1751.55580621173  525.444193788278   0.00171276   \n",
       "25           2   human            1049.5             297.5            0   \n",
       "26           3   human  1332.74039747032  370.650906877508   0.00600267   \n",
       "27           2   human  578.515593779322  177.484406220678   0.00864281   \n",
       "28           0   human               227                64   0.00440529   \n",
       "29           0   human  286.009963768116   76.990036231884   0.00349638   \n",
       "30           5   human   1145.6420481738  348.357951826201   0.00261862   \n",
       "31           3   human             520.5             154.5   0.00192123   \n",
       "32           2   human              1682  465.999999999999   0.00297265   \n",
       "33           4   human  1751.86491175721  525.135088242794   0.00399574   \n",
       "34           2   human  1050.11942257218  296.880577427822   0.00190455   \n",
       "35           2   human  1333.21212121212  370.787878787878  0.000750068   \n",
       "36           3   human  578.666666666666  177.333333333333            0   \n",
       "37           0   human  285.969782813975  77.0302171860245   0.00349687   \n",
       "38           1   human   1145.6600177305  348.339982269503            0   \n",
       "39           1   human  520.507554296507  154.492445703494            0   \n",
       "40           6   human  1681.97389375589  466.026106244114   0.00118908   \n",
       "41           3   human  1766.50000000001  504.499999999998  0.000566091   \n",
       "42           1   human            1047.5             296.5  0.000954654   \n",
       "43           1   human  1145.66666666667  348.333333333333  0.000872854   \n",
       "44           2   human  1682.98511904762   465.01488095238   0.00118836   \n",
       "45           1   human  1751.00000000001  525.999999999998            0   \n",
       "46           1   human  1332.12529832936  371.874701670644   0.00075068   \n",
       "47           2   human  578.648484848485  177.351515151515   0.00172816   \n",
       "48           0   human  226.990616621984  64.0093833780161   0.00881094   \n",
       "49           1   human             520.5             154.5            0   \n",
       "50           1   human  1682.16666666667  465.833333333332            0   \n",
       "51           0   human  1766.33993226257  504.660067737432   0.00169843   \n",
       "52           1   human  1751.19242020513  525.807579794871            0   \n",
       "53           0    duck  1332.33333333333  371.666666666666  0.000750563   \n",
       "54           0    duck  577.112612612612  178.806306306306   0.00173276   \n",
       "55           1    duck              1148  345.999999999999   0.00087108   \n",
       "56           1    duck  1766.16666666667  504.833333333332            0   \n",
       "57           1    duck            1049.5             297.5  0.000952835   \n",
       "58           0    duck  1330.33333333333  370.666666666666  0.000751691   \n",
       "59           0    duck  1146.46241134752  347.537588652482  0.000872248   \n",
       "60           2    duck  1681.73001508296  466.269984917043  0.000594626   \n",
       "61           0    duck   1766.1245210728  504.875478927201  0.000566211   \n",
       "62           4    duck  1751.33333333334  525.666666666665            0   \n",
       "63           0    duck  577.166666666667  178.833333333333    0.0017326   \n",
       "64           1    duck            1683.5  464.499999999999            0   \n",
       "65           2    duck  1766.99449944995  504.005500550053            0   \n",
       "66           1    duck              1334  369.999999999999            0   \n",
       "67           1    duck  285.985185185185  77.0148148148148   0.00349668   \n",
       "68           0    duck             520.5             154.5   0.00384246   \n",
       "69           1    duck              1682  465.999999999999            0   \n",
       "70           1    duck  1049.66666666666  297.333333333334            0   \n",
       "71           1    duck  1328.85441370224  372.145586297759  0.000752528   \n",
       "72           5    duck  1145.70741842189  348.292581578106   0.00261847   \n",
       "73           1    duck  1751.83333333334  525.166666666665            0   \n",
       "74           0   human                 0                 0            0   \n",
       "75           0   human                 0                 0            0   \n",
       "76           0   human                 0                 0            0   \n",
       "77           0   human                 0                 0            0   \n",
       "78           0   human                 0                 0            0   \n",
       "79           0   human                 0                 0            0   \n",
       "80           0   human                 0                 0            0   \n",
       "81           0   human                 0                 0            0   \n",
       "82           0   human                 0                 0            0   \n",
       "83           0   human                 0                 0            0   \n",
       "84           0   human                 0                 0            0   \n",
       "85           0   human                 0                 0            0   \n",
       "86           0   human                 0                 0            0   \n",
       "87           0   human                 0                 0            0   \n",
       "88           0   human                 0                 0            0   \n",
       "89           0   human                 0                 0            0   \n",
       "90           0   human                 0                 0            0   \n",
       "91           0   human                 0                 0            0   \n",
       "92           0   human                 0                 0            0   \n",
       "93           0   human                 0                 0            0   \n",
       "94           0   human                 0                 0            0   \n",
       "95           0   human                 0                 0            0   \n",
       "96           0   human                 0                 0            0   \n",
       "97           0   human                 0                 0            0   \n",
       "98           0   human                 0                 0            0   \n",
       "99           0   human                 0                 0            0   \n",
       "100          0    duck                 0                 0            0   \n",
       "101          0    duck                 0                 0            0   \n",
       "102          0    duck                 0                 0            0   \n",
       "103          0    duck                 0                 0            0   \n",
       "104          0    duck                 0                 0            0   \n",
       "105          0    duck                 0                 0            0   \n",
       "106          0    duck                 0                 0            0   \n",
       "107          0    duck                 0                 0            0   \n",
       "108          0    duck                 0                 0            0   \n",
       "109          0    duck                 0                 0            0   \n",
       "110          0    duck                 0                 0            0   \n",
       "111          0    duck                 0                 0            0   \n",
       "112          0    duck                 0                 0            0   \n",
       "113          0    duck                 0                 0            0   \n",
       "114          0    duck                 0                 0            0   \n",
       "115          0    duck                 0                 0            0   \n",
       "116          0    duck                 0                 0            0   \n",
       "117          0    duck                 0                 0            0   \n",
       "118          0    duck                 0                 0            0   \n",
       "119          0    duck                 0                 0            0   \n",
       "120          0    duck                 0                 0            0   \n",
       "121          0    duck                 0                 0            0   \n",
       "122          0    duck                 0                 0            0   \n",
       "123          0    duck                 0                 0            0   \n",
       "124          0    duck                 0                 0            0   \n",
       "125          0    duck                 0                 0            0   \n",
       "126          0    duck                 0                 0            0   \n",
       "127          0    duck                 0                 0            0   \n",
       "128          0    duck                 0                 0            0   \n",
       "\n",
       "     S_per_site  \n",
       "0    0.00269508  \n",
       "1             0  \n",
       "2    0.00578062  \n",
       "3    0.00215787  \n",
       "4    0.00795113  \n",
       "5    0.00772905  \n",
       "6     0.0167869  \n",
       "7    0.00538771  \n",
       "8             0  \n",
       "9    0.00577406  \n",
       "10    0.0021544  \n",
       "11   0.00192565  \n",
       "12   0.00672299  \n",
       "13   0.00270027  \n",
       "14   0.00289133  \n",
       "15   0.00861496  \n",
       "16   0.00397662  \n",
       "17            0  \n",
       "18            0  \n",
       "19            0  \n",
       "20            0  \n",
       "21   0.00650054  \n",
       "22   0.00429185  \n",
       "23   0.00991083  \n",
       "24            0  \n",
       "25   0.00672269  \n",
       "26   0.00809387  \n",
       "27    0.0112686  \n",
       "28            0  \n",
       "29            0  \n",
       "30    0.0143531  \n",
       "31    0.0194175  \n",
       "32   0.00429185  \n",
       "33   0.00761709  \n",
       "34   0.00673672  \n",
       "35   0.00539392  \n",
       "36    0.0169173  \n",
       "37            0  \n",
       "38   0.00287076  \n",
       "39   0.00647281  \n",
       "40    0.0128748  \n",
       "41   0.00594648  \n",
       "42   0.00337268  \n",
       "43   0.00287081  \n",
       "44   0.00430094  \n",
       "45   0.00190114  \n",
       "46   0.00268908  \n",
       "47     0.011277  \n",
       "48            0  \n",
       "49   0.00647249  \n",
       "50   0.00214669  \n",
       "51            0  \n",
       "52   0.00190184  \n",
       "53            0  \n",
       "54            0  \n",
       "55   0.00289017  \n",
       "56   0.00198085  \n",
       "57   0.00336134  \n",
       "58            0  \n",
       "59            0  \n",
       "60   0.00428936  \n",
       "61            0  \n",
       "62   0.00760938  \n",
       "63            0  \n",
       "64   0.00215285  \n",
       "65   0.00396821  \n",
       "66    0.0027027  \n",
       "67    0.0129845  \n",
       "68            0  \n",
       "69   0.00214592  \n",
       "70   0.00336323  \n",
       "71   0.00268712  \n",
       "72    0.0143557  \n",
       "73   0.00190416  \n",
       "74            0  \n",
       "75            0  \n",
       "76            0  \n",
       "77            0  \n",
       "78            0  \n",
       "79            0  \n",
       "80            0  \n",
       "81            0  \n",
       "82            0  \n",
       "83            0  \n",
       "84            0  \n",
       "85            0  \n",
       "86            0  \n",
       "87            0  \n",
       "88            0  \n",
       "89            0  \n",
       "90            0  \n",
       "91            0  \n",
       "92            0  \n",
       "93            0  \n",
       "94            0  \n",
       "95            0  \n",
       "96            0  \n",
       "97            0  \n",
       "98            0  \n",
       "99            0  \n",
       "100           0  \n",
       "101           0  \n",
       "102           0  \n",
       "103           0  \n",
       "104           0  \n",
       "105           0  \n",
       "106           0  \n",
       "107           0  \n",
       "108           0  \n",
       "109           0  \n",
       "110           0  \n",
       "111           0  \n",
       "112           0  \n",
       "113           0  \n",
       "114           0  \n",
       "115           0  \n",
       "116           0  \n",
       "117           0  \n",
       "118           0  \n",
       "119           0  \n",
       "120           0  \n",
       "121           0  \n",
       "122           0  \n",
       "123           0  \n",
       "124           0  \n",
       "125           0  \n",
       "126           0  \n",
       "127           0  \n",
       "128           0  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get rid of MP and NS columns and fill NA values with 0\n",
    "merged = merged[merged['gene'] != 'MP']\n",
    "merged = merged[merged['gene'] != 'NS']\n",
    "merged = merged.fillna(0.0)\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.000412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.001476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.002626</td>\n",
       "      <td>0.003258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.003370</td>\n",
       "      <td>0.002792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000693</td>\n",
       "      <td>0.000949</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species    variable      mean       std\n",
       "0   HA    duck  N_per_site  0.000451  0.000412\n",
       "1   HA    duck  S_per_site  0.001078  0.001476\n",
       "2   HA   human  N_per_site  0.002626  0.003258\n",
       "3   HA   human  S_per_site  0.003370  0.002792\n",
       "4   M1    duck  N_per_site  0.000693  0.000949"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a new dataframe with the mean and standard deviation per gene \n",
    "d = merged.melt(id_vars = ['gene','species'] ,value_vars = ['N_per_site','S_per_site'])\n",
    "d['value'] = pd.to_numeric(d['value'])\n",
    "mean = d.groupby([\"gene\", 'species','variable']).mean()\n",
    "mean.reset_index(inplace=True)\n",
    "mean.columns = ['gene','species','variable','mean']\n",
    "\n",
    "std = d.groupby([\"gene\", 'species','variable']).std()\n",
    "std.reset_index(inplace=True)\n",
    "std.columns = ['gene','species','variable','std']\n",
    "\n",
    "df1 = mean.merge(std, on=['gene','species','variable'], how=\"outer\")\n",
    "df1 = df1[df1['species'] != 0]\n",
    "df1.fillna(0)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.001476</td>\n",
       "      <td>duck synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.002626</td>\n",
       "      <td>0.003258</td>\n",
       "      <td>human nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.003370</td>\n",
       "      <td>0.002792</td>\n",
       "      <td>human synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000693</td>\n",
       "      <td>0.000949</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species    variable      mean       std                color\n",
       "0   HA    duck  N_per_site  0.000451  0.000412   duck nonsynonymous\n",
       "1   HA    duck  S_per_site  0.001078  0.001476      duck synonymous\n",
       "2   HA   human  N_per_site  0.002626  0.003258  human nonsynonymous\n",
       "3   HA   human  S_per_site  0.003370  0.002792     human synonymous\n",
       "4   M1    duck  N_per_site  0.000693  0.000949   duck nonsynonymous"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add in a color column \n",
    "df1['color'] = df1['species'] + \" \" + df1['variable']\n",
    "df1['color'] = df1['color'].str.replace(\"_per_site\",\"\")\n",
    "df1['color'] = df1['color'].str.replace(\"N\",\"nonsynonymous\")\n",
    "df1['color'] = df1['color'].str.replace(\"S\",\"synonymous\")\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000451</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.001078</td>\n",
       "      <td>0.001476</td>\n",
       "      <td>duck synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.002626</td>\n",
       "      <td>0.003258</td>\n",
       "      <td>human nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.003370</td>\n",
       "      <td>0.002792</td>\n",
       "      <td>human synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M1</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000693</td>\n",
       "      <td>0.000949</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species    variable      mean       std                color\n",
       "0   HA    duck  N_per_site  0.000451  0.000412   duck nonsynonymous\n",
       "1   HA    duck  S_per_site  0.001078  0.001476      duck synonymous\n",
       "2   HA   human  N_per_site  0.002626  0.003258  human nonsynonymous\n",
       "3   HA   human  S_per_site  0.003370  0.002792     human synonymous\n",
       "4   M1    duck  N_per_site  0.000693  0.000949   duck nonsynonymous"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output N/S ratio "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "            gene species   N_per_site  S_per_site     NoverS\n",
       "1             HA    duck 0.0004509564 0.001077965 0.41834065\n",
       "2             HA   human 0.0026260653 0.003369991 0.77924987\n",
       "3             M1    duck 0.0006930732 0.000000000        Inf\n",
       "4             M1   human 0.0016061623 0.004932866 0.32560429\n",
       "5             M2    duck 0.0000000000 0.000000000        NaN\n",
       "6             M2   human 0.0027513567 0.000000000        Inf\n",
       "7            NEP    duck 0.0006993369 0.002596903 0.26929652\n",
       "8            NEP   human 0.0008741568 0.000000000        Inf\n",
       "9  neuraminidase    duck 0.0001905669 0.001344915 0.14169445\n",
       "10 neuraminidase   human 0.0010722238 0.005042748 0.21262689\n",
       "11            NP    duck 0.0008723596 0.003449184 0.25291768\n",
       "12            NP   human 0.0010897117 0.004317580 0.25238948\n",
       "13           NS1    duck 0.0007684918 0.000000000        Inf\n",
       "14           NS1   human 0.0002401537 0.004857915 0.04943555\n",
       "15            PA    duck 0.0001189252 0.001717627 0.06923805\n",
       "16            PA   human 0.0015593991 0.005104169 0.30551477\n",
       "17           PB1    duck 0.0001132423 0.001189812 0.09517659\n",
       "18           PB1   human 0.0008893325 0.003969296 0.22405298\n",
       "19           PB2    duck 0.0000000000 0.001902708 0.00000000\n",
       "20           PB2   human 0.0011401469 0.002634346 0.43280077\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -w 1000 -h 500 -u px -i  df1  # this sets the size of the plot...otherwise, it will go off the page\n",
    "\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "require(reshape2)\n",
    "library(reshape2)\n",
    "\n",
    "x = dcast(df1, gene + species ~ variable, value.var=\"mean\")\n",
    "x$NoverS = x$N_per_site/x$S_per_site\n",
    "\n",
    "x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare nonsynonymous to synonymous diversity for full genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0013911435148309854,\n",
       " 0.002100978307636747,\n",
       " 0.003415974507324383,\n",
       " 0.00459111382375162,\n",
       " 0.40724645685972083)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# separate out human and duck \n",
    "df = merged.copy()\n",
    "\n",
    "human = df[df['species'] ==\"human\"]\n",
    "duck = df[df['species'] ==\"duck\"]\n",
    "\n",
    "meanNSh = human['N_per_site'].astype(float).mean()\n",
    "meanSh = human['S_per_site'].astype(float).mean()\n",
    "stdNSh = human['N_per_site'].astype(float).std()\n",
    "stdSh = human['S_per_site'].astype(float).std()\n",
    "NoverSh = meanNSh/meanSh\n",
    "\n",
    "meanNSh, stdNSh, meanSh, stdSh, NoverSh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.00039069523619551517,\n",
       " 0.0008691571147512671,\n",
       " 0.0013279114279839185,\n",
       " 0.00299946696552858,\n",
       " 0.2942178431197646)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meanNSd = duck['N_per_site'].astype(float).mean()\n",
    "meanSd = duck['S_per_site'].astype(float).mean()\n",
    "stdNSd = duck['N_per_site'].astype(float).std()\n",
    "stdSd = duck['S_per_site'].astype(float).std()\n",
    "NoverSd = meanNSd/meanSd\n",
    "\n",
    "meanNSd, stdNSd, meanSd, stdSd, NoverSd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>human</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>human nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.003416</td>\n",
       "      <td>0.004591</td>\n",
       "      <td>human synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>duck</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.001328</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>duck synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene species    variable      mean       std                color\n",
       "0  full\\ngenome   human  N_per_site  0.001391  0.002101  human nonsynonymous\n",
       "1  full\\ngenome   human  S_per_site  0.003416  0.004591     human synonymous\n",
       "2  full\\ngenome    duck  N_per_site  0.000391  0.000869   duck nonsynonymous\n",
       "3  full\\ngenome    duck  S_per_site  0.001328  0.002999      duck synonymous"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add in full genome values to plot \n",
    "fg = {'gene':['full\\ngenome','full\\ngenome','full\\ngenome','full\\ngenome'],\n",
    "      'species':['human','human','duck','duck'],\n",
    "      'variable':['N_per_site','S_per_site','N_per_site','S_per_site'],\n",
    "      'mean':[meanNSh,meanSh,meanNSd,meanSd],'std':[stdNSh,stdSh,stdNSd,stdSd],\n",
    "      'color':[\"human nonsynonymous\",\"human synonymous\",\"duck nonsynonymous\",\"duck synonymous\"]}\n",
    "fg_df = pd.DataFrame(fg)\n",
    "fg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>variable</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.005043</td>\n",
       "      <td>0.005676</td>\n",
       "      <td>human synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>human</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>human nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>human</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.003416</td>\n",
       "      <td>0.004591</td>\n",
       "      <td>human synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>duck</td>\n",
       "      <td>N_per_site</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>duck nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>full\\ngenome</td>\n",
       "      <td>duck</td>\n",
       "      <td>S_per_site</td>\n",
       "      <td>0.001328</td>\n",
       "      <td>0.002999</td>\n",
       "      <td>duck synonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene species    variable      mean       std                color\n",
       "39  neuraminidase   human  S_per_site  0.005043  0.005676     human synonymous\n",
       "0    full\\ngenome   human  N_per_site  0.001391  0.002101  human nonsynonymous\n",
       "1    full\\ngenome   human  S_per_site  0.003416  0.004591     human synonymous\n",
       "2    full\\ngenome    duck  N_per_site  0.000391  0.000869   duck nonsynonymous\n",
       "3    full\\ngenome    duck  S_per_site  0.001328  0.002999      duck synonymous"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df1.append(fg_df)\n",
    "df2.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py:191: FutureWarning: from_items is deprecated. Please use DataFrame.from_dict(dict(items), ...) instead. DataFrame.from_dict(OrderedDict(items)) may be used to preserve the key order.\n",
      "  res = PandasDataFrame.from_items(items)\n"
     ]
    }
   ],
   "source": [
    "%%R -w 1000 -h 500 -u px -i  df2,human_nonsyn_color,human_syn_color,duck_nonsyn_color,duck_syn_color # this sets the size of the plot...otherwise, it will go off the page\n",
    "\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "# this block of code is to ensure that even genes without SNPs get plotted so that the bars are all the same size\n",
    "species_list = unique(df2$species)\n",
    "gene_list = unique(df2$gene)\n",
    "variables_list = unique(df2$variable)\n",
    "\n",
    "for (species in species_list)\n",
    "{\n",
    "    for (gene in gene_list)\n",
    "    {\n",
    "        for (variable in variables_list)\n",
    "        {\n",
    "            x = df2[df2$species == species & df2$gene == gene & df2$variable == variable,]\n",
    "            if (nrow(x) == 0){\n",
    "                mean = 0\n",
    "                std = 0\n",
    "                if (variable == \"N_per_site\"){\n",
    "                    color = paste(species, \"nonsynonymous\")\n",
    "                } else{\n",
    "                    color = paste(species, \"synonymous\")\n",
    "                }\n",
    "                row_to_append = data.frame(gene, species, variable, mean, std, color)\n",
    "                df1 = rbind(df1, row_to_append)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "df2$gene = gsub(\"neuraminidase\",\"NA\", df2$gene)\n",
    "df2$colorf = factor(df2$color, levels=c(\"human nonsynonymous\",\"human synonymous\",\"duck nonsynonymous\",\"duck synonymous\"))\n",
    "df2$genef = factor(df2$gene, levels=c('PB2','PB1','PA','HA','NP','NA','M1','M2','NS1','NEP','full\\ngenome'))\n",
    "\n",
    "p <- ggplot(data=df2, aes(x=genef, y=mean, color=colorf, fill=colorf)) + \n",
    "    geom_col(position=\"dodge\")+   \n",
    "    geom_errorbar(data=df2, aes(x=genef, ymin=mean - std, ymax=mean+std), position=\"dodge\")+\n",
    "    labs(x=\"Gene\",y=\"SNVs/site\")+\n",
    "    scale_color_manual(values=c(human_nonsyn_color,human_syn_color,duck_nonsyn_color,duck_syn_color),breaks = c(\"human nonsynonymous\",\"human synonymous\",\"duck nonsynonymous\",\"duck synonymous\"),labels = c(\" human nonsynonymous\",\" human synonymous\",\" duck nonsynonymous\",\" duck synonymous\"))+\n",
    "    scale_fill_manual(values=c(human_nonsyn_color,human_syn_color,duck_nonsyn_color,duck_syn_color),breaks = c(\"human nonsynonymous\",\"human synonymous\",\"duck nonsynonymous\",\"duck synonymous\"),labels = c(\" human nonsynonymous\",\" human synonymous\",\" duck nonsynonymous\",\" duck synonymous\"))+\n",
    "    #ggtitle(\"random tables simulation\") + \n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    #facet_wrap(~genef, scales=\"free\")+\n",
    "    scale_y_continuous(limits=c(-0.005,0.015))+\n",
    "    theme(strip.text = element_text(size=16))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=16))+\n",
    "    theme(legend.text=element_text(size=16))+\n",
    "    theme(legend.title=element_blank())+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+\n",
    "    theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.65, \"cm\"))+\n",
    "    #theme(legend.key=element_rect(fill=NA,size = 2))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "ggsave(\"Fig-1c-NS_S_SNPs_per_site-2019-06-04.pdf\", p, width = 14, height = 6, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2019-06-04/individual-PDFs\")\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare human to duck"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=3.7550521007245345, pvalue=0.0002758219395960405)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare human to duck NS rates with a t-test\n",
    "h_NS = human['N_per_site'].astype(float)\n",
    "d_NS = duck['N_per_site'].astype(float)\n",
    "\n",
    "stats.ttest_ind(h_NS, d_NS, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=3.1240011827810825, pvalue=0.0022111998964308853)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare human to duck S rates with a t-test\n",
    "h_S = human['S_per_site'].astype(float)\n",
    "d_S = duck['S_per_site'].astype(float)\n",
    "\n",
    "stats.ttest_ind(h_S, d_S, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare nonsynonymous to synonymous within a species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=-3.5644838922096644, pvalue=0.0005419892842828632)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare NS to S for humans, full genome\n",
    "stats.ttest_ind(h_NS, h_S, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=-2.122133397749984, pvalue=0.038171341885881165)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare NS to S for ducks, full genome\n",
    "stats.ttest_ind(d_NS, d_S, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full genome results: \n",
    "\n",
    "**1. Test whether the rates of synonymous and nonsynonymous diversity are different across the full genome in each species separately:**\n",
    "\n",
    "Humans: p = 0.00054 (yes, more synonymous than nonsynonymous diversity)\n",
    "\n",
    "Ducks: p = 0.038 (yes, more synonymous than nonsynonymous diversity) \n",
    "\n",
    "\n",
    "**2. Test whether there is a difference between human and duck rates of nonsynonymous and synonymous diversity across the genomes**\n",
    "\n",
    "Human vs. duck nonsynonymous: p = 0.00012 (yes, humans have more nonsynonymous diversity than ducks) \n",
    "\n",
    "Human vs. duck synonymous: p = 0.00017 (yes, humans have more synonymous diversity than ducks)\n",
    "\n",
    "Overall, the mean, genome-wide ratio of piN/piS is human: 0.35, 0.29"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare diversity results per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>N_per_site</th>\n",
       "      <th>S_per_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>human</td>\n",
       "      <td>HA</td>\n",
       "      <td>0.009003</td>\n",
       "      <td>0.002695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>human</td>\n",
       "      <td>M1</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>human</td>\n",
       "      <td>NP</td>\n",
       "      <td>0.003484</td>\n",
       "      <td>0.005781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>human</td>\n",
       "      <td>PA</td>\n",
       "      <td>0.004749</td>\n",
       "      <td>0.002158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>human</td>\n",
       "      <td>PB1</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.007951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample species gene  N_per_site  S_per_site\n",
       "0  A_CAMBODIA_V0401301_2011   human   HA    0.009003    0.002695\n",
       "1  A_CAMBODIA_V0401301_2011   human   M1    0.000750    0.000000\n",
       "2  A_CAMBODIA_V0401301_2011   human   NP    0.003484    0.005781\n",
       "3  A_CAMBODIA_V0401301_2011   human   PA    0.004749    0.002158\n",
       "4  A_CAMBODIA_V0401301_2011   human  PB1    0.002263    0.007951"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# results per gene \n",
    "\n",
    "## This is the code that Dane helped me with \n",
    "merged['N_per_site'] = merged.N_per_site.astype(float)\n",
    "merged['S_per_site'] = merged.S_per_site.astype(float)\n",
    "\n",
    "human = merged[merged['species'] == 'human'][['sample','species','gene','N_per_site','S_per_site']]\n",
    "duck = merged[merged['species'] == 'duck'][['sample','species','gene','N_per_site','S_per_site']]\n",
    "\n",
    "human.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB2_NS</th>\n",
       "      <th>PB2_S</th>\n",
       "      <th>M2_NS</th>\n",
       "      <th>M2_S</th>\n",
       "      <th>PB1_NS</th>\n",
       "      <th>PB1_S</th>\n",
       "      <th>NEP_NS</th>\n",
       "      <th>NEP_S</th>\n",
       "      <th>HA_NS</th>\n",
       "      <th>HA_S</th>\n",
       "      <th>NP_NS</th>\n",
       "      <th>NP_S</th>\n",
       "      <th>NS1_NS</th>\n",
       "      <th>NS1_S</th>\n",
       "      <th>M1_NS</th>\n",
       "      <th>M1_S</th>\n",
       "      <th>PA_NS</th>\n",
       "      <th>PA_S</th>\n",
       "      <th>neuraminidase_NS</th>\n",
       "      <th>neuraminidase_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.001137</td>\n",
       "      <td>0.007729</td>\n",
       "      <td>0.004390</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.007951</td>\n",
       "      <td>0.003496</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009003</td>\n",
       "      <td>0.002695</td>\n",
       "      <td>0.003484</td>\n",
       "      <td>0.005781</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006501</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004749</td>\n",
       "      <td>0.002158</td>\n",
       "      <td>0.004766</td>\n",
       "      <td>0.016787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000569</td>\n",
       "      <td>0.001926</td>\n",
       "      <td>0.004405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.003497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003001</td>\n",
       "      <td>0.005388</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005774</td>\n",
       "      <td>0.001921</td>\n",
       "      <td>0.019417</td>\n",
       "      <td>0.001728</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001188</td>\n",
       "      <td>0.002154</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>0.006723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.001707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001698</td>\n",
       "      <td>0.009911</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002700</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>0.002891</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006473</td>\n",
       "      <td>0.008643</td>\n",
       "      <td>0.011269</td>\n",
       "      <td>0.001188</td>\n",
       "      <td>0.008615</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.001713</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008811</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.005946</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001501</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002619</td>\n",
       "      <td>0.014353</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006472</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.001905</td>\n",
       "      <td>0.006737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.003996</td>\n",
       "      <td>0.007617</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001698</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006003</td>\n",
       "      <td>0.008094</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001728</td>\n",
       "      <td>0.011277</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>0.004292</td>\n",
       "      <td>0.000955</td>\n",
       "      <td>0.003373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.005394</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.002871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.012875</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001902</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000751</td>\n",
       "      <td>0.002689</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001188</td>\n",
       "      <td>0.004301</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     PB2_NS     PB2_S     M2_NS  M2_S    PB1_NS     PB1_S    NEP_NS  NEP_S  \\\n",
       "0  0.001137  0.007729  0.004390   0.0  0.002263  0.007951  0.003496    0.0   \n",
       "1  0.000569  0.001926  0.004405   0.0  0.000000  0.003977  0.003497    0.0   \n",
       "2  0.001707  0.000000  0.004405   0.0  0.001698  0.009911  0.000000    0.0   \n",
       "3  0.001713  0.000000  0.008811   0.0  0.000566  0.005946  0.000000    0.0   \n",
       "4  0.003996  0.007617  0.000000   0.0  0.001698  0.000000  0.000000    0.0   \n",
       "5  0.000000  0.001901  0.000000   0.0  0.000000  0.000000  0.000000    0.0   \n",
       "6  0.000000  0.001902  0.000000   0.0  0.000000  0.000000  0.000000    0.0   \n",
       "7  0.000000  0.000000  0.000000   0.0       NaN       NaN  0.000000    0.0   \n",
       "\n",
       "      HA_NS      HA_S     NP_NS      NP_S    NS1_NS     NS1_S     M1_NS  \\\n",
       "0  0.009003  0.002695  0.003484  0.005781  0.000000  0.006501  0.000750   \n",
       "1  0.003001  0.005388  0.000000  0.005774  0.001921  0.019417  0.001728   \n",
       "2  0.000000  0.002700  0.001742  0.002891  0.000000  0.006473  0.008643   \n",
       "3  0.001501  0.000000  0.002619  0.014353  0.000000  0.006472  0.000000   \n",
       "4  0.006003  0.008094  0.000000  0.002871  0.000000  0.000000  0.001728   \n",
       "5  0.000750  0.005394  0.000873  0.002871  0.000000  0.000000  0.000000   \n",
       "6  0.000751  0.002689  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "7  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "       M1_S     PA_NS      PA_S  neuraminidase_NS  neuraminidase_S  \n",
       "0  0.000000  0.004749  0.002158          0.004766         0.016787  \n",
       "1  0.000000  0.001188  0.002154          0.000953         0.006723  \n",
       "2  0.011269  0.001188  0.008615          0.000000         0.006723  \n",
       "3  0.016917  0.000000  0.004292          0.001905         0.006737  \n",
       "4  0.011277  0.002973  0.004292          0.000955         0.003373  \n",
       "5  0.000000  0.001189  0.012875          0.000000         0.000000  \n",
       "6  0.000000  0.001188  0.004301          0.000000         0.000000  \n",
       "7  0.000000  0.000000  0.002147          0.000000         0.000000  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "humanList = []\n",
    "geneList2 = []\n",
    "\n",
    "geneList = list(set(human['gene']))\n",
    "for g in geneList:\n",
    "    geneList2.append(g +\"_\"+ \"NS\")\n",
    "    geneList2.append(g +\"_\"+ \"S\")\n",
    "\n",
    "for gene in geneList:\n",
    "    genedf_NS = human[human['gene'] == gene]['N_per_site']\n",
    "    genedf_NS = genedf_NS.reset_index(drop=True)\n",
    "    \n",
    "    genedf_S = human[human['gene'] == gene]['S_per_site']\n",
    "    genedf_S = genedf_S.reset_index(drop=True)\n",
    "\n",
    "    humanList.append(genedf_NS)\n",
    "    humanList.append(genedf_S)\n",
    "    \n",
    "    human_compare = pd.concat(humanList, axis=1)\n",
    "\n",
    "human_compare.columns = geneList2\n",
    "human_compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.5268435255733626, pvalue=0.17063681720420376)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare per gene rates with t-tests \n",
    "stats.ttest_rel(human_compare['PB2_NS'], human_compare['PB2_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.2171449254243947, pvalue=0.06846646867778151)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['PB1_NS'].dropna(), human_compare['PB1_S'].dropna(), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.297541019055976, pvalue=0.055190722764687394)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['PA_NS'], human_compare['PA_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-0.6212746601013918, pvalue=0.5541033360863585)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['HA_NS'], human_compare['HA_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.340520767088095, pvalue=0.05180523784536771)"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['NP_NS'], human_compare['NP_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.6462458749001097, pvalue=0.03312170880337058)"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['neuraminidase_NS'], human_compare['neuraminidase_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.439018297267753, pvalue=0.19331889683798115)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['M1_NS'], human_compare['M1_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=2.3750332882913616, pvalue=0.049240392614373935)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['M2_NS'], human_compare['M2_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.136614837816369, pvalue=0.06998326345647055)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['NS1_NS'], human_compare['NS1_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=1.5275252266264807, pvalue=0.17047066201139502)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(human_compare['NEP_NS'], human_compare['NEP_S'], axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Repeat for ducks: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PB2_NS</th>\n",
       "      <th>PB2_S</th>\n",
       "      <th>M2_NS</th>\n",
       "      <th>M2_S</th>\n",
       "      <th>PB1_NS</th>\n",
       "      <th>PB1_S</th>\n",
       "      <th>NEP_NS</th>\n",
       "      <th>NEP_S</th>\n",
       "      <th>HA_NS</th>\n",
       "      <th>HA_S</th>\n",
       "      <th>NP_NS</th>\n",
       "      <th>NP_S</th>\n",
       "      <th>NS1_NS</th>\n",
       "      <th>NS1_S</th>\n",
       "      <th>M1_NS</th>\n",
       "      <th>M1_S</th>\n",
       "      <th>PA_NS</th>\n",
       "      <th>PA_S</th>\n",
       "      <th>neuraminidase_NS</th>\n",
       "      <th>neuraminidase_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001981</td>\n",
       "      <td>0.003497</td>\n",
       "      <td>0.012985</td>\n",
       "      <td>0.000751</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.002890</td>\n",
       "      <td>0.003842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000595</td>\n",
       "      <td>0.004289</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>0.003361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000566</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000752</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000872</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001733</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002153</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003968</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002703</td>\n",
       "      <td>0.002618</td>\n",
       "      <td>0.014356</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002146</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000753</td>\n",
       "      <td>0.002687</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PB2_NS     PB2_S  M2_NS  M2_S    PB1_NS     PB1_S    NEP_NS     NEP_S  \\\n",
       "0     0.0  0.007609    0.0   0.0  0.000000  0.001981  0.003497  0.012985   \n",
       "1     0.0  0.001904    0.0   0.0  0.000566  0.000000  0.000000  0.000000   \n",
       "2     0.0  0.000000    0.0   0.0  0.000000  0.003968  0.000000  0.000000   \n",
       "3     0.0  0.000000    0.0   0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "4     0.0  0.000000    0.0   0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "      HA_NS      HA_S     NP_NS      NP_S    NS1_NS  NS1_S     M1_NS  M1_S  \\\n",
       "0  0.000751  0.000000  0.000871  0.002890  0.003842    0.0  0.001733   0.0   \n",
       "1  0.000752  0.000000  0.000872  0.000000  0.000000    0.0  0.001733   0.0   \n",
       "2  0.000000  0.002703  0.002618  0.014356  0.000000    0.0  0.000000   0.0   \n",
       "3  0.000753  0.002687  0.000000  0.000000  0.000000    0.0  0.000000   0.0   \n",
       "4  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  0.000000   0.0   \n",
       "\n",
       "      PA_NS      PA_S  neuraminidase_NS  neuraminidase_S  \n",
       "0  0.000595  0.004289          0.000953         0.003361  \n",
       "1  0.000000  0.002153          0.000000         0.003363  \n",
       "2  0.000000  0.002146          0.000000         0.000000  \n",
       "3  0.000000  0.000000          0.000000         0.000000  \n",
       "4  0.000000  0.000000          0.000000         0.000000  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now do duck\n",
    "duckList = []\n",
    "geneList2 = []\n",
    "\n",
    "geneList = list(set(duck['gene']))\n",
    "for g in geneList:\n",
    "    geneList2.append(g +\"_\"+ \"NS\")\n",
    "    geneList2.append(g +\"_\"+ \"S\")\n",
    "\n",
    "for gene in geneList:\n",
    "    genedf_NS = duck[duck['gene'] == gene]['N_per_site']\n",
    "    genedf_NS = genedf_NS.reset_index(drop=True)\n",
    "    \n",
    "    genedf_S = duck[duck['gene'] == gene]['S_per_site']\n",
    "    genedf_S = genedf_S.reset_index(drop=True)\n",
    "\n",
    "    duckList.append(genedf_NS)\n",
    "    duckList.append(genedf_S)\n",
    "    \n",
    "    duck_compare = pd.concat(duckList, axis=1)\n",
    "\n",
    "duck_compare.columns = geneList2\n",
    "duck_compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.2912402353854437, pvalue=0.26618746843730634)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compare per gene rates with t-tests \n",
    "stats.ttest_rel(duck_compare['PB2_NS'], duck_compare['PB2_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.2778315723962415, pvalue=0.27042850836356574)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['PB1_NS'], duck_compare['PB1_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-2.2483967603499666, pvalue=0.08780096016018495)"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['PA_NS'], duck_compare['PA_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-0.8775596794688688, pvalue=0.4297310535859451)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['HA_NS'], duck_compare['HA_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.1017649691596973, pvalue=0.33240060439985797)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['NP_NS'], duck_compare['NP_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-1.5969859963595219, pvalue=0.18550664214899515)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['neuraminidase_NS'], duck_compare['neuraminidase_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=1.632993158871184, pvalue=0.17780780898035436)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['M1_NS'], duck_compare['M1_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=nan, pvalue=nan)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['M2_NS'].dropna(), duck_compare['M2_S'].dropna(), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=1.0000000000000002, pvalue=0.3739009663000589)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['NS1_NS'], duck_compare['NS1_S'], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_relResult(statistic=-0.9999999999999999, pvalue=0.373900966300059)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ttest_rel(duck_compare['NEP_NS'], duck_compare['NEP_S'], axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results: \n",
    "\n",
    "\n",
    "For humans, only NP and NA have any evidence that the mean nonsynonymous diversity is different from the mean synonymous diversity. \n",
    "\n",
    "For ducks, no genes have any evidence that the mean synonymous diversity is different from the mean nonsynonymous diversity. \n",
    "\n",
    "\n",
    "When you combine the full genomes across samples, there is evidence that there is more synonymous than nonsynonymous diversity in humans, but not in ducks. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare and plot ratios of NS/S between humans and ducks per gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>sample</th>\n",
       "      <th>gene</th>\n",
       "      <th>nonsynonymous</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>species</th>\n",
       "      <th>Nsites</th>\n",
       "      <th>Ssites</th>\n",
       "      <th>N_per_site</th>\n",
       "      <th>S_per_site</th>\n",
       "      <th>N_over_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>HA</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.009003</td>\n",
       "      <td>0.002695</td>\n",
       "      <td>3.340370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>1332.9534942423</td>\n",
       "      <td>371.046505757698</td>\n",
       "      <td>0.000750</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>750.213720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>NP</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>human</td>\n",
       "      <td>1148.01647108678</td>\n",
       "      <td>345.983528913223</td>\n",
       "      <td>0.003484</td>\n",
       "      <td>0.005781</td>\n",
       "      <td>0.602750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PA</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>1684.40017392789</td>\n",
       "      <td>463.421018125091</td>\n",
       "      <td>0.004749</td>\n",
       "      <td>0.002158</td>\n",
       "      <td>2.201002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A_CAMBODIA_V0401301_2011</td>\n",
       "      <td>PB1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>human</td>\n",
       "      <td>1767.9268235821</td>\n",
       "      <td>503.073176417901</td>\n",
       "      <td>0.002263</td>\n",
       "      <td>0.007951</td>\n",
       "      <td>0.284555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index                    sample gene nonsynonymous synonymous species  \\\n",
       "0      0  A_CAMBODIA_V0401301_2011   HA            12          1   human   \n",
       "1      1  A_CAMBODIA_V0401301_2011   M1             1          0   human   \n",
       "2      2  A_CAMBODIA_V0401301_2011   NP             4          2   human   \n",
       "3      3  A_CAMBODIA_V0401301_2011   PA             8          1   human   \n",
       "4      4  A_CAMBODIA_V0401301_2011  PB1             4          4   human   \n",
       "\n",
       "             Nsites            Ssites  N_per_site  S_per_site    N_over_S  \n",
       "0   1332.9534942423  371.046505757698    0.009003    0.002695    3.340370  \n",
       "1   1332.9534942423  371.046505757698    0.000750    0.000001  750.213720  \n",
       "2  1148.01647108678  345.983528913223    0.003484    0.005781    0.602750  \n",
       "3  1684.40017392789  463.421018125091    0.004749    0.002158    2.201002  \n",
       "4   1767.9268235821  503.073176417901    0.002263    0.007951    0.284555  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first, make sure everything is numeric\n",
    "merged2 = merged.copy()\n",
    "merged2['N_per_site'] = pd.to_numeric(merged2['N_per_site'])\n",
    "merged2['S_per_site'] = pd.to_numeric(merged2['S_per_site'])\n",
    "\n",
    "# add a pseudocount to all 0 values. Instead of having a diversity of 0, have it be very low so that I don't get a \n",
    "# division by 0 error \n",
    "merged2['N_per_site'] = merged2['N_per_site'].replace(0, 0.000001)\n",
    "merged2['S_per_site'] = merged2['S_per_site'].replace(0, 0.000001)\n",
    "merged2['N_over_S'] = merged2['N_per_site']/merged2['S_per_site']\n",
    "merged2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>N_over_S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>3.340370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M1</td>\n",
       "      <td>human</td>\n",
       "      <td>750.213720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP</td>\n",
       "      <td>human</td>\n",
       "      <td>0.602750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA</td>\n",
       "      <td>human</td>\n",
       "      <td>2.201002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PB1</td>\n",
       "      <td>human</td>\n",
       "      <td>0.284555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene species    N_over_S\n",
       "0   HA   human    3.340370\n",
       "1   M1   human  750.213720\n",
       "2   NP   human    0.602750\n",
       "3   PA   human    2.201002\n",
       "4  PB1   human    0.284555"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take the mean values of N over S for each gene for birds and humans \n",
    "x = merged2[['gene','species','N_over_S']]\n",
    "human = x[x['species'] == 'human']\n",
    "duck = x[x['species'] == 'duck']\n",
    "human.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA</td>\n",
       "      <td>duck</td>\n",
       "      <td>300.706930</td>\n",
       "      <td>411.175847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HA</td>\n",
       "      <td>human</td>\n",
       "      <td>282.144509</td>\n",
       "      <td>795.925644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M1</td>\n",
       "      <td>duck</td>\n",
       "      <td>693.673172</td>\n",
       "      <td>948.481805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M1</td>\n",
       "      <td>human</td>\n",
       "      <td>310.261401</td>\n",
       "      <td>630.089922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M2</td>\n",
       "      <td>duck</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>M2</td>\n",
       "      <td>human</td>\n",
       "      <td>2201.239787</td>\n",
       "      <td>3328.340001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NEP</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.853859</td>\n",
       "      <td>0.326781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NEP</td>\n",
       "      <td>human</td>\n",
       "      <td>874.906831</td>\n",
       "      <td>1618.161024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NP</td>\n",
       "      <td>duck</td>\n",
       "      <td>174.946438</td>\n",
       "      <td>389.803832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NP</td>\n",
       "      <td>human</td>\n",
       "      <td>0.342557</td>\n",
       "      <td>0.327777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NS1</td>\n",
       "      <td>duck</td>\n",
       "      <td>769.291835</td>\n",
       "      <td>1717.952769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NS1</td>\n",
       "      <td>human</td>\n",
       "      <td>0.512426</td>\n",
       "      <td>0.522241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PA</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.427912</td>\n",
       "      <td>0.525280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PA</td>\n",
       "      <td>human</td>\n",
       "      <td>0.493931</td>\n",
       "      <td>0.727250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PB1</td>\n",
       "      <td>duck</td>\n",
       "      <td>113.642450</td>\n",
       "      <td>252.994279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PB1</td>\n",
       "      <td>human</td>\n",
       "      <td>404.754639</td>\n",
       "      <td>1069.746813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PB2</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.600131</td>\n",
       "      <td>0.547543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PB2</td>\n",
       "      <td>human</td>\n",
       "      <td>213.711497</td>\n",
       "      <td>603.379336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>duck</td>\n",
       "      <td>0.656753</td>\n",
       "      <td>0.480555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>neuraminidase</td>\n",
       "      <td>human</td>\n",
       "      <td>119.221231</td>\n",
       "      <td>336.346617</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene species         mean          std\n",
       "0              HA    duck   300.706930   411.175847\n",
       "1              HA   human   282.144509   795.925644\n",
       "2              M1    duck   693.673172   948.481805\n",
       "3              M1   human   310.261401   630.089922\n",
       "4              M2    duck     1.000000     0.000000\n",
       "5              M2   human  2201.239787  3328.340001\n",
       "6             NEP    duck     0.853859     0.326781\n",
       "7             NEP   human   874.906831  1618.161024\n",
       "8              NP    duck   174.946438   389.803832\n",
       "9              NP   human     0.342557     0.327777\n",
       "10            NS1    duck   769.291835  1717.952769\n",
       "11            NS1   human     0.512426     0.522241\n",
       "12             PA    duck     0.427912     0.525280\n",
       "13             PA   human     0.493931     0.727250\n",
       "14            PB1    duck   113.642450   252.994279\n",
       "15            PB1   human   404.754639  1069.746813\n",
       "16            PB2    duck     0.600131     0.547543\n",
       "17            PB2   human   213.711497   603.379336\n",
       "18  neuraminidase    duck     0.656753     0.480555\n",
       "19  neuraminidase   human   119.221231   336.346617"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make mean and standard deviation dataframes for plotting\n",
    "mean = x.groupby([\"gene\", 'species']).mean()\n",
    "std = x.groupby([\"gene\", 'species']).std()\n",
    "mean = mean.reset_index()\n",
    "std = std.reset_index()\n",
    "NoverS_df = mean.merge(std, on=['gene','species'])\n",
    "NoverS_df.columns = ['gene','species','mean','std']\n",
    "NoverS_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA+gAAAH0CAYAAACuKActAAAEDWlDQ1BJQ0MgUHJvZmlsZQAAOI2N\nVV1oHFUUPrtzZyMkzlNsNIV0qD8NJQ2TVjShtLp/3d02bpZJNtoi6GT27s6Yyc44M7v9oU9FUHwx\n6psUxL+3gCAo9Q/bPrQvlQol2tQgKD60+INQ6Ium65k7M5lpurHeZe58853vnnvuuWfvBei5qliW\nkRQBFpquLRcy4nOHj4g9K5CEh6AXBqFXUR0rXalMAjZPC3e1W99Dwntf2dXd/p+tt0YdFSBxH2Kz\n5qgLiI8B8KdVy3YBevqRHz/qWh72Yui3MUDEL3q44WPXw3M+fo1pZuQs4tOIBVVTaoiXEI/MxfhG\nDPsxsNZfoE1q66ro5aJim3XdoLFw72H+n23BaIXzbcOnz5mfPoTvYVz7KzUl5+FRxEuqkp9G/Aji\na219thzg25abkRE/BpDc3pqvphHvRFys2weqvp+krbWKIX7nhDbzLOItiM8358pTwdirqpPFnMF2\nxLc1WvLyOwTAibpbmvHHcvttU57y5+XqNZrLe3lE/Pq8eUj2fXKfOe3pfOjzhJYtB/yll5SDFcSD\niH+hRkH25+L+sdxKEAMZahrlSX8ukqMOWy/jXW2m6M9LDBc31B9LFuv6gVKg/0Szi3KAr1kGq1GM\njU/aLbnq6/lRxc4XfJ98hTargX++DbMJBSiYMIe9Ck1YAxFkKEAG3xbYaKmDDgYyFK0UGYpfoWYX\nG+fAPPI6tJnNwb7ClP7IyF+D+bjOtCpkhz6CFrIa/I6sFtNl8auFXGMTP34sNwI/JhkgEtmDz14y\nSfaRcTIBInmKPE32kxyyE2Tv+thKbEVePDfW/byMM1Kmm0XdObS7oGD/MypMXFPXrCwOtoYjyyn7\nBV29/MZfsVzpLDdRtuIZnbpXzvlf+ev8MvYr/Gqk4H/kV/G3csdazLuyTMPsbFhzd1UabQbjFvDR\nmcWJxR3zcfHkVw9GfpbJmeev9F08WW8uDkaslwX6avlWGU6NRKz0g/SHtCy9J30o/ca9zX3Kfc19\nzn3BXQKRO8ud477hLnAfc1/G9mrzGlrfexZ5GLdn6ZZrrEohI2wVHhZywjbhUWEy8icMCGNCUdiB\nlq3r+xafL549HQ5jH+an+1y+LlYBifuxAvRN/lVVVOlwlCkdVm9NOL5BE4wkQ2SMlDZU97hX86Ei\nlU/lUmkQUztTE6mx1EEPh7OmdqBtAvv8HdWpbrJS6tJj3n0CWdM6busNzRV3S9KTYhqvNiqWmuro\niKgYhshMjmhTh9ptWhsF7970j/SbMrsPE1suR5z7DMC+P/Hs+y7ijrQAlhyAgccjbhjPygfeBTjz\nhNqy28EdkUh8C+DU9+z2v/oyeH791OncxHOs5y2AtTc7nb/f73TWPkD/qwBnjX8BoJ98VVBg/m8A\nAEAASURBVHgB7d0JnBxVnTjw30wSEoFAQoKbcBNhIQZ2IxEUUIgIqyhyBVDwYtUAfhZXdJfDdb1Q\nF1mNy7Xkn12CRrxwxaiQrAsqmAUCGuU0BCKHCBJymJAQEnLM/POK7XZ6Mkf3MD1T0/Wtz6fTdbyq\neu/7Kj31q3r1qql1yxAGAgQIECBAgAABAgQIECBAoF8Fmvt173ZOgAABAgQIECBAgAABAgQIZAIC\ndAcCAQIECBAgQIAAAQIECBDIgYAAPQeVIAsECBAgQIAAAQIECBAgQECA7hggQIAAAQIECBAgQIAA\nAQI5EBCg56ASZIEAAQIECBAgQIAAAQIECAjQHQMECBAgQIAAAQIECBAgQCAHAgL0HFSCLBAgQIAA\nAQIECBAgQIAAAQG6Y4AAAQIECBAgQIAAAQIECORAQICeg0qQBQIECBAgQIAAAQIECBAgIEB3DBAg\nQIAAAQIECBAgQIAAgRwICNBzUAmyQIAAAQIECBAgQIAAAQIEBOiOAQIECBAgQIAAAQIECBAgkAMB\nAXoOKkEWCBAgQIAAAQIECBAgQICAAN0xQIAAAQIECBAgQIAAAQIEciAgQM9BJcgCAQIECBAgQIAA\nAQIECBAQoDsGCBAgQIAAAQIECBAgQIBADgQE6DmoBFkgQIAAAQIECBAgQIAAAQICdMcAAQIECBAg\nQIAAAQIECBDIgYAAPQeVIAsECBAgQIAAAQIECBAgQECA7hggQIAAAQIECBAgQIAAAQI5EBCg56AS\nZIEAAQIECBAgQIAAAQIECAjQHQMECBAgQIAAAQIECBAgQCAHAgL0HFSCLBAgQIAAAQIECBAgQIAA\nAQG6Y4AAAQIECBAgQIAAAQIECORAQICeg0qQBQIECBAgQIAAAQIECBAgIEB3DBAgQIAAAQIECBAg\nQIAAgRwICNDrUAktLS2RPgYCBAgQIECAAAECBAgQIFCtgAC9Wqka0l199dUxe/bsGtaQlAABAgQI\nECBAgAABAgSKLiBAL/oRoPwECBAgQIAAAQIECBAgkAsBAXouqkEmCBAgQIAAAQIECBAgQKDoAgL0\noh8Byk+AAAECBAgQIECAAAECuRAQoOeiGmSCAAECBAgQIECAAAECBIouIEAv+hGg/AQIECBAgAAB\nAgQIECCQCwEBei6qQSYIECBAgAABAgQIECBAoOgCAvSiHwHKT4AAAQIECBAgQIAAAQK5EBCg56Ia\nZIIAAQIECBAgQIAAAQIEii4gQC/6EaD8BAgQIECAAAECBAgQIJALAQF6LqpBJggQIECAAAECBAgQ\nIECg6AIC9KIfAcpPgAABAgQIECBAgAABArkQEKDnohpkggABAgQIECBAgAABAgSKLiBAL/oRoPwE\nCBAgQIAAAQIECBAgkAsBAXouqkEmCBAgQIAAAQIECBAgQKDoAgL0oh8Byk+AAAECBAgQIECAAAEC\nuRAQoOeiGmSCAAECBAgQIECAAAECBIouIEAv+hGg/AQIECBAgAABAgQIECCQC4HBuciFTBAgQIAA\nAQIECBAgUBZYcPv8uHve/5an8zTSFE3xNye9I/YZv3+esiUvBBpCQIDeENWoEAQIECBAgAABAo0k\nMPf7P4i7bp2XyyI1NTXFiJ1GCtBzWTsyNdAFNHEf6DUo/wQIECBAgAABAgT6UCAF6AYCBOojIECv\nj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQI\nECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVx\nSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECA\nAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6ut\nEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBA\ngACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyA\nAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAEC\nBOojIECvj6utEiBAgAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBA\ngAABAgQIECBAgACBmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACB\nmgQE6DVxSUyAAAECBAgQIECAAAECBOojIECvj6utEiBAgAABAgQIECBAgACBmgQaIkBvaWnpttCt\nra19lqbbHUlAgAABAgQIECBAgAABAgTaCQzoAP073/lOHHDAATF06NDYaaed4p3vfGc89dRTFUWc\nNWtWTJ48Obbddts45JBD4rbbbqtYniZ6K81WGzaDAAECBAgQIECAAAECBAhUKTBgA/Sbbropzjjj\njHjta18bP/3pT+PSSy+NBQsWxNve9rbYsGFDVvx58+bFWWedFaecckrMnz8/Jk2aFMcee2zcf//9\nZZ7eSlPeoBECBAgQIECAAAECBAgQINADgaYtTb+7b/vdgw3Xe5UUaC9atCgee+yxaGpqynZ3/fXX\nx7ve9a74xS9+EUcccUS8+tWvjoMOOii++c1vlrNz4IEHZnfSZ86cmc3rrTTlHWwZueqqq2Ls2LEx\nZcqUtrONEyBAgAABAgQIEKhK4OLzzo+7bp1XVdq+TtTc3BxnnP3BOOOcD/X1ru2PQMMLDNg76FOn\nTo0ZM2aUg/NUU3vvvXdWYc8991zW1P2hhx6Kk046qaISTzjhhJg7d242LzWH7400FTswQYAAAQIE\nCBAgQIAAAQIEeiAwuAfr5GKVk08+eat8fOtb34rBgwdnzd7T3fU07LrrrhXp0vSyZcsidSy3ePHi\nXkmTriIaCBAgQIAAAQIECBAgQIDAyxFomMjyjjvuyO6of/zjH8+al69ZsyZzGTVqVIXPyJEjY/Pm\nzbFixYrorTSlHXz729+OadOmxd13312a5ZsAAQIECBAgQIAAAQIECFQl0BAB+l133RXveMc74vWv\nf3187nOfywqe7qSnofR8ejbR5p/UkVxvpSltNt2VT8H/AH2sv1QM3wQIECBAgAABAgQIECDQDwID\ntol7yernP/95HH/88VnHbz/+8Y9j2LBh2aLUSVsaVq1alX2X/lm5cmU2usMOO2R32tPEy01T2vZ7\n3vOebDR1EmcgQIAAAQIECBAgQIAAAQK1CAzoO+i33nprHHfccXH00UdnHb9tv/325bKPGTMmG3/m\nmWfK89LIkiVLYvTo0TF8+PDorTQVOzBBgAABAgQIECBAgAABAgR6IDBgA/Tf/va3WXB+4oknxg03\n3FC+c14ySHfQJ0yYEHPmzCnNyr7T+9OPOuqobLy30lTswAQBAgQIECBAgAABAgQIEOiBwIBt4n7e\needlz3ofeuihce2111YUPb0Dfb/99otzzz03Pvaxj8UxxxyTff793/89Fi5cGP/1X/9VTt9bacob\nNEKAAAECBAgQIECAAAECBHogMCAD9NRs/ac//WlW3L//+7/fqtjXXHNNFqCnd6WnV6mdfvrpsXHj\nxhg/fnzW0/u4cePK6/RWmvIGjRAgQIAAAQIECBAgQIAAgR4ING3pcby1B+sNqFXWr18fS5cujT32\n2KPTfPdWmrSD1Elcaj4/ZcqUTvdnAQECBAgQIECAAIHOBC4+7/y469Z5nS3u1/nNzc1xxtkfjDPO\n+VC/5sPOCTSiwIB9Br2Wykg9u3cVnKdt9VaaWvIlLQECBAgQIECAAAECBAgQKAkUIkAvFdY3AQIE\nCBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAAAQIECBAgUCgBAXqhqlthCRAgQIAAAQIECBAgQCCv\nAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWqboUlQIAAAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIE\nCBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAAAQIE8iogQM9rzcgXAQIECBAgQIAAAQIECBRKQIBe\nqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAAAQIECBAgUCgBAXqhqlthCRAgQIAAAQIE\nCBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWqboUlQIAAAQIECBAgQIAAgbwKCNDzWjPy\nRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAAAQIE8iogQM9rzcgXAQIECBAgQIAAAQIE\nCBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAAAQIECBAgUCgBAXqhqlthCRAg\nQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWqboUlQIAAAQIECBAgQIAAgbwK\nCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAAAQIE8iogQM9rzcgXAQIECBAg\nQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAAAQIECBAgUCgBAXqh\nqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWqboUlQIAAAQIECBAg\nQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAAAQIE8iogQM9rzcgX\nAQIECBAgQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAAAQIECBAg\nUCgBAXqhqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWqboUlQIAA\nAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAAAQIE8iog\nQM9rzcgXAQIECBAgQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18ECBAgQIAA\nAQIECBAgUCgBAXqhqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBAoQQE6IWq\nboUlQIAAAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIECBAgQIAA\nAQIE8iogQM9rzcgXAQIECBAgQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AAPa81I18E\nCBAgQIAAAQIECBAgUCgBAXqhqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIECBAgQIBA\noQQE6IWqboUlQIAAAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6FZYAAQIE\nCBAgQIAAAQIE8iogQM9rzcgXAQIECBAgQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIECBDIq4AA\nPa81I18ECBAgQIAAAQIECBAgUCgBAXqhqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEgQIAAAQIE\nCBAgQIBAoQQE6IWqboUlQIAAAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKFEhCgF6q6\nFZYAAQIECBAgQIAAAQIE8iogQM9rzcgXAQIECBAgQIAAAQIECBRKQIBeqOpWWAIECBAgQIAAAQIE\nCBDIq4AAPa81I18ECBAgQIAAAQIECBAgUCgBAXqhqlthCRAgQIAAAQIECBAgQCCvAgL0vNaMfBEg\nQIAAAQIECBAgQIBAoQQE6IWqboUlQIAAAQIECBAgQIAAgbwKCNDzWjPyRYAAAQIECBAgQIAAAQKF\nEmiYAH3Dhg2xefPml1V5ra2t3a5fTZpuNyIBAQIECBAgQIAAAQIECBBoJ9AQAfrjjz8eu+yyS/zk\nJz+pKN7zzz8f48ePj/3337/iM3Xq1Ip0s2bNismTJ8e2224bhxxySNx2220Vy9NENWm2WskMAgQI\nECBAgAABAgQIECBQpcCAD9BTcH788cfHihUrtirygw8+GIsWLYpjjz02TjvttPLnjW98YzntvHnz\n4qyzzopTTjkl5s+fH5MmTcrS33///TWlKSc2QoAAAQIECBAgQIAAAQIEeiAwuAfr5GaV6dOnx/nn\nnx+77bZbh3lKQfb2228f06ZNi+bmjq9FnHPOOXHqqafGueeem20jbfP222+Pyy+/PGbOnJnNqyZN\nhxkwkwABAgQIECBAgAABAgQIVCnQcdRa5cr9neyLX/xifOQjH4m5c+d2mJX77rsvJk6cmAXnLS0t\nW6V56qmn4qGHHoqTTjqpYtkJJ5xQ3mY1aSpWNkGAAAECBAgQIECAAAECBHogMKAD9HvuuScuueSS\n2GabbToserqDPmjQoDjzzDNjxIgRsfvuu8ell14apWB98eLF2Xq77rprxfppetmyZVm6atJUrGyC\nAAECBAgQIECAAAECBAj0QGBAB+g777xzl0VOAfqCBQti7733jssuuyz23HPPuOiii+ILX/hCtt6a\nNWuy71GjRlVsZ+TIkVmP8Om59mrSlFZ+//vfHwcddFBce+21pVm+CRAgQIAAAQIECBAgQIBAVQID\n+hn0rkq4adOmSM+T77vvvnHwwQdnST/wgQ/Em970puyue3p2ffDgl4rf1NTU4abSq9uqSVNa+Stf\n+Uqkdb7xjW+UZvkmQIAAAQIECBAgQIAAAQJVCTRsgJ4C6zPOOGMrhNQhXHqN2sMPPxxjx47Nlq9a\ntaoi3cqVK7PpHXbYoao0pZVLd/SHDx9emuWbAAECBAgQIECAAAECBAhUJTCgm7h3VcLUNP3OO++M\n1atXVyQbNmxYNj1kyJAYM2ZMNv7MM89UpFmyZEmMHj06UqBdTZqKlU0QIECAAAECBAgQIECAAIEe\nCDRsgP7kk0/G4YcfHldccUUFyw033BA77rhj7LPPPtnd8QkTJsScOXMq0tx0001x1FFHZfPSXfbu\n0lSsbIIAAQIECBAgQIAAAQIECPRAoGED9BRUH3rooXHllVfG7Nmzs87evv71r8ctt9ySdRQ3dOjQ\njCu9/3zWrFmRAvd0tz31Cr9w4cLsu+RZTZpSWt8ECBAgQIAAAQIECBAgQKAnAg37DHrCuP766+OD\nH/xgnHzyyZlNarL++c9/Pi688MKy1dSpUyO9Su3000+PjRs3xvjx42PGjBkxbty4mtKUExshQIAA\nAQIECBAgQIAAAQI9EGhq3TL0YL0BtUrqBG758uVZ0N3c3HGjgfXr18fSpUtjjz326LRs1aRJK191\n1VVZ8/kpU6Z0ui0LCBAgQIAAAQIECHQmcPF558ddt87rbHG/zk/n02ec/cE445wP9Ws+7JxAIwo0\n9B30UoWNGDEi0qerIXUe11VwntatJk1X+7CMAAECBAgQIECAAAECBAh0JtDx7eTOUptPgAABAgQI\nECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVA\ngF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIE\nCBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQ\na/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBA\ngAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4X\nVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAg\nQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OS\nmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAAB\nAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhsl\nQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAA\nAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgAB\nAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI\n1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAA\nAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1\nCQjQa/OSmgABAgQIECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQI\nECBAgAABAgQI1EVAgF4XVhslQIAAAQIECBAgQIAAAQK1CQjQa/OSmgABAgQIECBAgAABAgQI1EVA\ngF4XVhslQIAAAQIECBAgQIAAAQK1CfRLgP6b3/wm0qc0PPHEE3HeeefFEUccEZ/61Kfi2WefLS3y\nTYAAAQIECBAgQIAAAQIECiHQpwF6S0tLHH/88TFp0qSYPXt2Brx27do47rjjYubMmTF8+PCYPn16\nNr1u3bpCVIBCEiBAgAABAgQIECBAgACBJNCnAfq0adPilltuiS996UtxwQUXZDWQ5i1cuDBuvPHG\nmDNnTixYsCAeeeSRuPLKK9UQAQIECBAgQIAAAQIECBAojMDgvizpT37ykzj66KPjwgsvLO/2Rz/6\nUbzuda+LyZMnZ/P22muveOtb3xp33313OY0RAgQIECBAgAABAgQIECDQ6AJ9egf9gQceiDe/+c1l\n05UrV8a9994bxxxzTHleGtlzzz3j0UcfrZhnggABAgQIECBAgAABAgQINLJAnwboI0aMiKeffrrs\n+bOf/SzSc+ltg/a0cNGiRbHvvvuW0xkhQIAAAQIECBAgQIAAAQKNLtCnAfrEiROzZ9BffPHFzPW6\n666LUaNGxWGHHVZ2/v3vfx/z5s2LAw44oDzPCAECBAgQIECAAAECBAgQaHSBPn0G/TOf+Uz2vHm6\nOz569Oi455574uqrr44hQ4bExo0b4/rrr886kNtuu+1i6tSpjW6vfAQIECBAgAABAgQIECBAoCzQ\np3fQJ0yYEKmjuCOPPDJSc/cZM2bEhz/84Swzzz33XLz3ve+NQYMGReo4bpdddiln0ggBAgQIECBA\ngAABAgQIEGh0gT69g55epZZ6a09N29sPqan7448/HqkXdwMBAgQIECBAgAABAgQIECiaQJ/eQf/s\nZz+bNW1PQXp6F3pq4t7a2pqZNzU1Cc6LdvQpLwECBAgQIECAAAECBAiUBfo0QP/1r38dTzzxRJx5\n5plx3333Za9XGzt2bLz//e+Pb3/727F8+fJyxowQIECAAAECBAgQIECAAIEiCfRpgJ5gU0CeAvTv\nfOc7sXTp0ux583HjxsVVV12VPXd+8MEHx6c+9am4++67i1QPykqAAAECBAgQIECAAAECBRfo8wC9\nrXdzc3PWq3vq3f3OO++MZ599Ns4///zsXenp20CAAAECBAgQIECAAAECBIoi0KedxHWHOnLkyDjt\ntNOyT3dpLSdAgAABAgQIECBAgAABAo0k0C930H/zm99E+pSG9Fz6eeedF0cccUTWvD3dSTcQIECA\nAAECBAgQIECAAIEiCfRpgN7S0hLHH398TJo0KWbPnp05r127No477riYOXNmDB8+PKZPn55Nr1u3\nrkj1oKwECBAgQIAAAQIECBAgUHCBPg3Qp02bFrfcckv2irULLrggo0/zFi5cGOkd6XPmzIkFCxbE\nI488EldeeWXBq0bxCRAgQIAAAQIECBAgQKBIAn36DPpPfvKTOProo+PCCy8sG//oRz/KOopL70ZP\nw1577RVvfetb9eKeafiHAAECBAgQIECAAAECBIoi0Kd30B944IF485vfXLZduXJl3Hvvvdn70Msz\nt4zsueee8eijj7adZZwAAQIECBAgQIAAAQIECDS0QJ8G6CNGjMheoVYS/dnPfhbpufS2QXtatmjR\noth3331LyXwTIECAAAECBAgQIECAAIGGF+jTAH3ixInZM+gvvvhiBnvdddfFqFGj4rDDDitD//73\nv4958+bFAQccUJ5nhAABAgQIECBAgAABAgQINLpAnz6D/pnPfCZ73jzdHR89enTcc889cfXVV8eQ\nIUNi48aNcf3112cdyG233XYxderURrdXPgIECBAgQIAAAQIECBAgUBbo0zvoEyZMiNRR3JFHHhmp\nufuMGTPiwx/+cJaZ5557Lt773vfGoEGDInUct8suu5QzaYQAAQIECBAgQIAAAQIECDS6QJ/eQU+Y\nb3jDG7JPe9jU1P3xxx/PenFvv8w0AQIECBAgQIAAAQIECBBodIE+DdBTj+2rV6/u0vTJJ5/Mlu+8\n884xfvz4LtNaSIAAAQIECBAgQIAAAQIEGkWgTwP0s88+O375y19WZXfqqafG9773varSSkSAAAEC\nBAgQIECAAAECBAa6QJ8G6F/72tdi7dq1nZr94Ac/iC9/+csxbNiwOOGEEzpN19GCDRs2ZM+vp2fY\nOxpaW1ujqampo0Xleb2VprxBIwQIECBAgAABAgQIECBAoEqBPu0k7tWvfnUcfPDBW31e9apXxWWX\nXZb14J6eUb///vvj3e9+d5VFiOzZ9dSpXOqArv0wa9asmDx5cmy77bZxyCGHxG233dY+SfRWmq02\nbAYBAgQIECBAgAABAgQIEKhSoE8D9I7ylHpsT727//CHP4wrrrgibr311hg3blxHSTuclzqWO/74\n42PFihVbLU/vUz/rrLPilFNOifnz58ekSZPi2GOPzS4AlBL3VprS9nwTIECAAAECBAgQIECAAIGe\nCPRbgL5y5cp4z3veEyeeeGKk96Knu+Yf+chHum2G3raQ06dPjwMPPDB7h3rb+aXxc845J9Kz7Oee\ne25MnDgxUvp99tknLr/88lKS6K005Q0aIUCAAAECBAgQIECAAAECPRDolwD9xhtvzO6az549O2va\n/otf/CJSM/dahy9+8YtZUD937tytVn3qqafioYceipNOOqliWXq2vZS+t9JU7MAEAQIECBAgQIAA\nAQIECBDogUCfdhKX7pp/9KMfjeuuuy57F3rqNC7d0e7pcM8990R6HVsKtNsPixcvzmbtuuuuFYvS\n9LJly6KlpSV6K01zc79c56golwkCBAgQIECAAAECBAgQGNgCfRqgn3zyyVknba94xSuywPqiiy7q\nVO+www6Lj3/8450uTwtScN7ZsGbNmmzRqFGjKpKMHDkyNm/enD2z3ltpSvlIz7OnixCPPPJIjB07\ntmK/JggQIECAAAECBAgQIECAQFcCfRqgp87f1q9fn+XnmWee6SpfHXb61uUK7RYOHvxS0Tp7tVp6\nLVtvpSnt+o477ojHHnssnn766TjyyCNLs30TIECAAAECBAgQIECAAIFuBfo0QJ85c2a3GeqtBKU7\n2KtWrarYZLrDnYYddtihfJf75aYp7eATn/hENnrVVVeVZvkmQIAAAQIECBAgQIAAAQJVCTTsw9Nj\nxozJANrfqV+yZEmMHj06hg8fHr2VpippiQgQIECAAAECBAgQIECAQBcCDRugpzvo6f3qc+bMqSj+\nTTfdFEcddVQ2r7fSVOzABAECBAgQIECAAAECBAgQ6IFAwwboySK9/3zWrFlxww03xOrVq+OSSy6J\nhQsXZt8lq95KU9qebwIECBAgQIAAAQIECBAg0BOBPn0GvScZfDnrTJ06NXuV2umnnx4bN26M8ePH\nx4wZMyJ1VlcaeitNaXu+CRAgQIAAAQIECBAgQIBATwSaWrcMPVlxIK2Teo5funRp7LHHHp1mu7fS\npB2kTuJS8/kpU6Z0uj8LCBAgQIAAAQIECHQmcPF558ddt87rbHG/zm9ubo4zzv5gnHHOh/o1H3ZO\noBEFGrqJe6nChg0b1mVwntL1VprSPn0TIECAAAECBAgQIECAAIFaBAoRoNcCIi0BAgQIECBAgAAB\nAgQIEOgPAQF6f6jbJwECBAgQIECAAAECBAgQaCcgQG8HYpIAAQIECBAgQIAAAQIECPSHgAC9P9Tt\nkwABAgQIECBAgAABAgQItBMQoLcDMUmAAAECBAgQIECAAAECBPpDQIDeH+r2SYAAAQIECBAgQIAA\nAQIE2gkI0NuBmCRAgAABAgQIECBAgAABAv0hIEDvD3X7JECAAAECBAgQIECAAAEC7QQE6O1ATBIg\nQIAAAQIECBAgQIAAgf4QEKD3h7p9EiBAgAABAgQIECBAgACBdgIC9HYgJgkQIECAAAECBAgQIECA\nQH8ICND7Q90+CRAgQIAAAQIECBAgQIBAO4HB7aZNEiBAgAABAgQIECBAgACBmgWeeerpWLtmTc3r\n9cUKQ4ZsE3vuM64vdvWy9iFAf1l8ViZAgAABAgQIECBAgACBJHD2ie+MQYNSI+2m3IFsePHFuObG\n78fY3XfLXd7aZkiA3lbDOAECBAgQIECAAAECBAj0SGDz5s2xaePGHq1b75W2GbpNbMxp3tqW3TPo\nbTWMEyBAgAABAgQIECBAgACBfhIQoPcTvN0SIECAAAECBAgQIECAAIG2AgL0thrGCRAgQIAAAQIE\nCBAgQIBAPwkI0PsJ3m4JECBAgAABAgQIECBAgEBbAQF6Ww3jBAgQIECAAAECBAgQIECgnwQE6P0E\nb7cECBAgQIAAAQIECBAgQKCtgAC9rYZxAgQIECBAgAABAgQIECDQTwIC9H6Ct1sCBAgQIECAAAEC\nBAgQINBWQIDeVsM4AQIECBAgQIAAAQIECBDoJwEBej/B2y0BAgQIECBAgAABAgQIEGgrIEBvq2Gc\nAAECBAgQIECAAAECBAj0k4AAvZ/g7ZYAAQIECBAgQIAAAQIECLQVEKC31TBOgAABAgQIECBAgAAB\nAgT6SUCA3k/wdkuAAAECBAgQIECAAAECBNoKCNDbahgnQIAAAQIECBAgQIAAAQL9JDC4n/ZrtwQI\nECBAgAABAgQINKjA0yufjTXr1uaydM3NzfGqV+4eg5oH5TJ/MlVsAQF6setf6QkQIECAAAECBAj0\nusAPfnVzbNi8KZqjqde3/XI3uKllU7zr9W+P3UeNfbmbsj6BXhcQoPc6qQ0SIECAAAECBAgQKLZA\nS2trbNoSoOdx2GbQkGjdkj8DgTwKeAY9j7UiTwQIECBAgAABAgQIECBQOAEBeuGqXIEJECBAgAAB\nAgQIECBAII8CAvQ81oo8ESBAgAABAgQIECBAgEDhBATohatyBSZAgAABAgQIECBAgACBPAoI0PNY\nK/JEgAABAgQIECBAgAABAoUTEKAXrsoVmAABAgQIECBAgAABAgTyKCBAz2OtyBMBAgQIECBAgAAB\nAgQIFE5AgF64KldgAgQIECBAgAABAgQIEMijgAA9j7UiTwQIECBAgAABAgQIECBQOAEBeuGqXIEJ\nECBAgAABAgQIECBAII8Cg/OYKXkiQIAAAQIECBB4SeCPK5fGw888lluOfcfsFbvtNCa3+ZMxAgQI\nDCQBAfpAqi15JUCAAAECBAonsPDp38Wvn/htbsu9fuMGAXpua0fGCBAYaAKauA+0GpNfAgQIECBA\ngAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6\nQ1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIE\nCBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0\nGpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAAB\nAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1ar\nQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAg\nQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNf\nAgQIECBAgAABAgQIEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQI\nEGhIAQF6Q1arQhEgQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6Q1arQhEg\nQIAAAQIECBAgQIDAQBMQoA+0GpNfAgQIECBAgAABAgQIEGhIAQF6m2ptbW1tM9XxaDVpOl7TXAIE\nCBAgQIAAAQIECBAg0LlAQwfozz//fIwfPz7233//is/UqVMrRGbNmhWTJ0+ObbfdNg455JC47bbb\nKpaniWrSbLWSGQQIECBAgAABAgQIECBAoEqBhg7QH3zwwVi0aFEce+yxcdppp5U/b3zjG8s88+bN\ni7POOitOOeWUmD9/fkyaNClLf//999eUppzYCAECBAgQIECAAAECBAgQ6IHA4B6sM2BWSUH29ttv\nH9OmTYvm5o6vRZxzzjlx6qmnxrnnnpuVa/r06XH77bfH5ZdfHjNnzszmVZNmwKDIKAECBAgQIECA\nAAECBAjkUqDjqDWXWa09U/fdd19MnDgxC85bWlq22sBTTz0VDz30UJx00kkVy0444YSYO3duNq+a\nNBUrmyBAgAABAgQIECBAgAABAj0QaOgAPd1BHzRoUJx55pkxYsSI2H333ePSSy+NUrC+ePHijGzX\nXXetoEvTy5Yty9JVk6ZiZRMECBAgQIAAAQIECBAgQKAHAg0foC9YsCD23nvvuOyyy2LPPfeMiy66\nKL7whS9kVGvWrMm+R40aVUE3cuTI2Lx5c6xYsSKqSVNa+ZOf/GS8853vjNmzZ5dm+SZAgAABAgQI\nECBAgAABAlUJNOwz6Js2bYr0PPm+++4bBx98cIbxgQ98IN70pjfFJZdcEueff34MHvxS8ZuamjrE\n2rBhQ1VpSitPmTIlC+jnzJlTmuWbAAECBAgQIECAAAECBAhUJdCwAXoKvs8444ytEFKHcOk1ag8/\n/HCMHTs2W75q1aqKdCtXrsymd9hhh6rSlFY+6KCDstEHHnigNMs3AQIECBAgQIAAAQIECBCoSqBh\nm7inpul33nlnrF69ugJi2LBh2fSQIUNizJgx2fgzzzxTkWbJkiUxevToGD58eFVpKlY2QYAAAQIE\nCBAgQIAAAQIEeiDQsAH6k08+GYcffnhcccUVFSw33HBD7LjjjrHPPvtkd8cnTJgQ7Zuk33TTTXHU\nUUdl66W77N2lqdiBCQIECBAgQIAAAQIECBAg0AOBhg3QU1B96KGHxpVXXpl12pbuqH/961+PW265\nJesobujQoRlXev/5rFmzIgXu6W57ej594cKF2XfJs5o0pbS+CRAgQIAAAQIECBAgQIBATwQa9hn0\nhHH99dfHBz/4wTj55JMzm9Rk/fOf/3xceOGFZaupU6dGepXa6aefHhs3bozx48fHjBkzYty4cTWl\nKSc2QoAAAQIECBAgQIAAAQIEeiDQ0AF6eu/5zTffHKkTuOXLl2dBd3NzZaOB9J70adOmxRe/+MVY\nunRp7LHHHlsxVpNmq5XMIECAAAECBAgQIECAAAECNQg0dIBechgxYkSkT1dD6jyuo+C87TrVpGmb\n3jgBAgQIECBAgAABAgQIEKhWoPJ2crVrSUeAAAECBAgQIECAAAECBAj0qoAAvVc5bYwAAQIECBAg\nQIAAAQIECPRMQIDeMzdrESBAgAABAgQIECBAgACBXhUQoPcqp40RIECAAAECBAgQIECAAIGeCQjQ\ne+ZmLQIECBAgQIAAAQIECBAg0KsCAvRe5bQxAgQIECBAgAABAgQIECDQMwEBes/crEWAAAECBAgQ\nIECAAAECBHpVoBDvQe9VMRsj0IsCP5/z33Hb3Jt7cYu9t6mmpohT/vZ9ceCk1/TeRm2JAAECBAgQ\nIECAAIFOBQTondJYQKD+Arff/PNYcPud9d9RD/bQ1NwU+x0wQYDeAzurECBAgAABAgQIEOiJgCbu\nPVGzDoECCDSlW+gGAgQIECBAgAABAgT6TECA3mfUdkSAAAECBAgQIECAAAECBDoXEKB3bmMJAQIE\nCBAgQIAAAQIECBDoMwEBep9R2xEBAgQIECBAgAABAgQIEOhcQIDeuY0lBAgQIECAAAECBAgQIECg\nzwQE6H1GbUcECBAgQIAAAQIECBAgQKBzAQF65zaWECBAgAABAgQIECBAgACBPhMQoPcZtR0RIECA\nAAECBAgQIECAAIHOBQTondtYQoAAAQIECBAgQIAAAQIE+kxAgN5n1HZEgAABAgQIECBAgAABAgQ6\nFxCgd25jCQECBAgQIECAAAECBAgQ6DMBAXqfUdsRAQIECBAgQIAAAQIECBDoXECA3rmNJQQIECBA\ngAABAgQIECBAoM8EBOh9Rm1HBAgQIECAAAECBAgQIECgc4HBnS+yhAABAgQIECBAgEDXAgtunx+3\nT/+vrhP109LJx74ljjnhuH7au90SIECgdgEBeu1m1iBAgEDdBTZu3hSbWzbXfT892cGg5kExZJA/\nHz2xsw6BRhT407Jl8fD8X+ayaIMHDxGg57JmZIoAgc4EnGF1JmM+AQIE+klg05bg/LKfzIqW1pZ+\nykHXu21qaoqPvuV9MXTwNl0ntJQAgWIIbPlNMBAgQIBA7wh4Br13HG2FAAECvSawuaUlt8F5KmRr\na2ts3pzPu/u9Vgk2RIAAAQIECBDoBwEBej+g2yUBAgQIECBAgAABAgQIEGgvoIl7exHTBAgQINCl\nQLqDfuaxJ8amF17sMl1/LNxm6NCY/oNvx8hRo/pj9/ZJgAABAgQIEHhZAgL0l8VnZQIECBRT4MV1\n62Pjlk/ehs1bnt9f89xqAXreKkZ+CBAgQIAAgaoENHGvikkiAgQIEBgIAk2hs6qBUE/ySIAAAQIE\nCHQsIEDv2MVcAgQIECBAgAABAgQIECDQpwIC9D7ltjMCBAgQIECAAAECBAgQINCxgAC9YxdzCRAg\nQIAAAQIECBAgQIBAnwoI0PuU284IECBAgAABAgQIECBAgEDHAgL0jl3MJUCAAAECBAgQIECAAAEC\nfSogQO9TbjsjQIAAAQIECBAgQIAAAQIdCwjQO3YxlwABAgQIECBAgAABAgQI9KmAAL1Pue2MAAEC\nBAgQIECAAAECBAh0LDC449nmEvizwMoVK+L+X/76zzNyNvaq8fvFbnvtmbNcyQ4BAgQIECBAgAAB\nAgRqExCg1+ZVyNQ3z74xvvX/rokhQ4bkrvybNm2KSYe9Pj59+ZdzlzcZIkCAAAECBAgQIECAQC0C\nAvRatAqatrW1NdJn3Qsv5FKgpWVzLvMlUwQIECBAgAABAgQIEKhFwDPotWhJS4AAAQIECBAgQIAA\nAQIE6iQgQK8TrM0SIECAAAECBAgQIECAAIFaBATotWhJS4AAAQIECBAgQIAAAQIE6iQgQK8TrM0S\nIECAAAECBAgQIECAAIFaBHQSV4uWtAQIECAwoAVWr3s+bvjVzbFp86ZclmP08JFx0muPyWXeZIoA\nAQIECBCov4AAvf7G9kCAAAECORFYs35trHh+VWzO6dsfUv4MBAgQIECAQHEFNHEvbt0rOQECBAop\n0NzUVMhyKzQBAgQIECCQfwEBev7rSA4JECBAgAABAgQIECBAoAACAvQCVLIiEiBAgAABAgQIECBA\ngED+BQTo+a8jOSRAgAABAgQIECBAgACBAggI0AtQyYpIgAABAgQIECBAgAABAvkXEKDnv47kkAAB\nAgQIECBAgAABAgQKICBAL0AlKyIBAgQIECBAgAABAgQI5F/Ae9DzX0dySIAAAQIECLwMgfUvrIv5\nt82L1paWl7GV+q26yx67xf5/dWD9dmDLBAgQIDBgBAToA6aqZLSnAk+vfDaefW55T1ev63rNu+0Y\nQ7Z/RWx8fl1d92PjBAgQKLLAPXf/Mv7t0xfH4CFDcsfQsrkldho9Kq6dOzt3eZMhAgQIEOh7AQF6\n35vbYx8L3Lrw7liyalk0NTX18Z67393g/XaOnSfsFX+8+6HuE0tBgAABAj0SaG1tjSHbbBPr1r7Q\no/XrvVJLTu/s17vctk+AAAECWwsI0Lc2MafBBNKJ2ebWLc0aW3NYsDzmKYdMskSAAAECBAgQIECg\nCAI6iStCLSsjAQIECBAgQIAAAQIECOReQICe+yqSQQIECBAgQIAAAQIECBAogoAAvQi1rIwECBAg\nQIAAAQIECBAgkHsBAXruq0gGCRAgQIAAAQIECBAgQKAIAgL0ItSyMhIgQIAAAQIECBAgQIBA7gUE\n6LmvIhkkQIAAAQIECBAgQIAAgSIICNCLUMvKSIAAAQIECBAgQIAAAQK5FxCg576KZJAAAQIECBAg\nQIAAAQIEiiAgQC9CLSsjAQIECBAgQIAAAQIECOReQICe+yqSQQIECBAgQIAAAQIECBAogoAAvQi1\nrIwECBAgQIAAAQIECBAgkHsBAXqNVdTa2lrjGpITIECAAAECBAgQIECAAIHuBQTo3RtlKWbNmhWT\nJ0+ObbfdNg455JC47bbbqlxTMgIECBAgQIAAAQIECBAg0L2AAL17o5g3b16cddZZccopp8T8+fNj\n0qRJceyxx8b9999fxdqSECBAgAABAgQIECBAgACB7gUE6N0bxTnnnBOnnnpqnHvuuTFx4sSYPn16\n7LPPPnH55ZdXsbYkBAgQIECAAAECBAgQIECgewEBejdGTz31VDz00ENx0kknVaQ84YQTYu7cuRXz\nTBAgQIAAAQIECBAgQIAAgZ4KDO7pikVZb/HixVlRd91114oip+lly5ZFS0tLNDe/dJ3j8ccfj3Xr\n1sXy5ctj7NixFem7m1j74guxZNXy7pL1y/IXhrbEkO2Gxebnnu+X/dtpPwlU0R/isjV/irXr1/VT\nBrve7fJnn41Y9WI0NTV1nbAfljYPao4DDpoYQ7bZph/23ti7bKmiI888d/aZ/qbc/6tf57KShu+4\nQ+z9l/t2mrcXN22IP6x4Jqqogk63Ua8FKzY9H9uM3D7WrX2hXrso9nbzWOlV1sjS51ZE+luWx6F5\nzPAY/IqhsWndi3nM3oDO0+aWzfHwgwtj2ZAnc1eOpuamOHDSa7o8R3hs6R8i/ebmcRj5l7vFnxbl\nzzVZtbRUcXKbA1QBejeVsGbNmizFqFGjKlKOHDkyNm/eHCtWrIidd945W/aVr3wlu9uegvQDDzyw\nIn13E7979sm445HfdJesX5a/uFNT7HLQfrHi3kf7Zf/d7XTcfn/ZZZIxI0bHyhdWd5mmvxa+uOVY\nad7QEjuOHNFfWehiv02x6557dLE8Yt6iBfHsc/m8sLRm1XPxyPW/iI3P5e+EPAXo//zVL8U+4/fv\n0HfwoEGx03Y7xrqN+TwpW718ZWy33bbROnRYh/nvz5lDhw2LHUd0/v9pu6GviFdsMyw2bTk5y9uQ\nLhysfHJJTLvic3nLWpafUX/xyvjqN67pNG/PbrnIfPMDd3S6vD8XbNq8MXY7dELET/N5Qrvvq8d3\nyTNq+MjsuO0yUT8t3LhhQ2xavT6nf8ciXrV/1+cIDzz1SCz642P9pNf1bofuPybG7L9XrHliywXn\n3A1bzg333L3LXI3ZcXQ8u3pFl2n6a+G659fG9bOuiQ2r1vZXFjrd70vnCJd2eo7Q0toSty68O9Zv\nyuc5wv7vODwWLn8+Wje3dFrG/lqQzhF22HHH/tp91ftt2nJCMDAuJVRdpN5NmJqxv/3tb490Jz09\nd14avvvd78bpp58eqQl8+7vrV111VXYHfcqUKaXkvgkQIECAAAECBAgQIECAQJcCnkHvkifKTdVX\nrVpVkXLlypXZ9A477FAx3wQBAgQIECBAgAABAgQIEOiJgAC9G7UxY8ZkKZ555pmKlEuWLInRo0fH\n8OHDK+abIECAAAECBAgQIECAAAECPREQoHejljp7mzBhQsyZM6ci5U033RRHHXVUxTwTBAgQIECA\nAAECBAgQIECgpwIC9Crk0vvPZ82aFTfccEOsXr06Lrnkkli4cGH2XcXqkhAgQIAAAQIECBAgQIAA\ngW4F9OLeLVHE1KlTs07iUqdwGzdujPHjx8eMGTNi3LhxVawtCQECBAgQIECAAAECBAgQ6F5AL+7d\nG5VTrF+/PpYuXRp77NH1q6f04l4mM0KAAAECBAgQIECAAAECVQpo4l4lVEo2bMu787oLzmvYnKQE\nCBAgQIAAAQIECBAgQKAsIEAvUxghQIAAAQIECBAgQIAAAQL9JyBA7z97eyZAgAABAgQIECBAgAAB\nAmUBAXqZwggBAgQIECBAgAABAgQIEOg/AQF6/9nbMwECBAgQIECAAAECBAgQKAsI0MsURggQIECA\nAAECBAgQIECAQP8JCND7z96eCRAgQIAAAQIECBAgQIBAWUCAXqYwQoAAAQIECBAgQIAAAQIE+k9A\ngN5/9vZMgAABAgQIECBAgAABAgTKAgL0MoURAgQIECBAgAABAgQIECDQfwKD+2/XjbvnP/7xj3Hf\nfffFs88+2xCFbGlpidbW1hg0aFBDlCdPhdi8eXM0NTVFc7NrZb1dL5s2bcqO2eRr6F2BZDt4sD8f\nvasa2e9s+k1g29uyEenvWPqwrY+tc4Ted01bdI5QH9e0VecI9bVttN/ac845p0/P1Z1h1eH4fPOb\n3xy/+93vYqeddqrD1vt+k7fddlusXr06jj/++L7feYPv8Xvf+17sueee8brXva7BS9r3xbvqqqvi\nbW97W4wbN67vd97ge/z0pz8dF154YWy33XYNXtK+Ld7y5cvjmmuuiX/6p3/q2x0XYG8LFy6MO++8\nMz70oQ8VoLR9W0TnCPXzTucIe+yxR7z+9a+v304KumXnCPWr+HSOcMEFF8T2229fv500+JYF6HWo\n4BSgp0+jDOmkcenSpXHqqac2SpFyU4758+fHa17zGrZ1qJHrrrsuJk+e7OJHHWw/85nPZBfsGuUi\nZB2IerTJJ554ImbNmuX3oEd6Xa90yy23xOLFi9l2zdSjpekcIbUYdI7QI74uV7rrrrucI3Qp1POF\nzhF6btfdmp/97Gezc4RRo0Z1l9TyTgQE6J3AmP1ngb322itGjBjx5xnGek3ggAMOiN12263XtmdD\nfxY45JBDYuTIkX+eYazXBN70pjfFNtts02vbs6GXBLbddtt4wxvegKMOAq985SuzQKcOmy78Jp0j\n1O8QmDBhQuy+++7120GBt+wcoX6Vn26OOEd4eb5NW54ban15m7A2AQIECBAgQIAAAQIECBAg8HIF\n9Ez1cgWtT6DOAukaWurMxECAAAECBAaKgL9dA6WmipnPdHymxzdT55FdDRs3buxqsWUE6iIgQK8L\na/43+v3vfz/rPTz1cJ0+qSlKamr9iU98opz51MFWaXmn00nQAAAQTElEQVTqZTx1Znb22WdnvYqW\nEv3qV7+Kd7/73VlT4re//e0xZ86c0qLCfr/zne8suyW/1JFWes78pptuykyef/75iuVDhw6NAw88\nML71rW91aJZ6jrz44os7XFb0maXjM32nYzQ1aT/hhBPiqaeeqqDZsGFD1mljI/UNUVHAXp4488wz\nY+rUqVtt9eMf/3iccsopFfPZVnB0OzFlypTYZ599Yv369RVpU8dwp59+ejavmt/nipVNZCfZ6Xfg\nYx/72FYa++67b/zgBz+omD937tzsd3jGjBkV801sLZCO1WS7ww47RPr/3nZI0zvuuGPsv//+bWdn\n4/52bUWy1YxvfOMbme173/verZbdeOON2bJPfepT5WXXXnttHHHEEbHzzjvH3/3d32UdEpcXFnwk\nBdrV/gakTszScfsXf/EXWUdm6Ty2/bGdOFPfFX/9139daNlqXdMxnPw7+pTeatV2WVfnbIUG/7/C\nC9ALfBSkDp7+53/+J/ukk5f3ve99cemll8b1119fVkknjCnNj3/84/jHf/zHbNkll1xSXn7eeefF\nkCFDInVkcvjhh0cKTtetW1deXtSRE088sWyb/gCPHj060o/XihUryiTTp0/P0qTA/Mgjj4z3v//9\n8dvf/ra8PI187Wtfi//4j/+omGeiUuCrX/1q5phOuNN46ngvHZdth7QsXYC6/fbbI3XEZeg9Aba1\nWz766KPR9ne0oy1U8/vc0XpFn3fllVdmrzntzuGb3/xmHHTQQZECHkN1AilQTz22tx1uvfXWWLNm\nTdtZ2bi/XVuRdDojnUOl39H0SrW2ww9/+MMs2CnNSxeeP/KRj2Q3RVI9PP744/HJT36ytNj3/wl0\n9xvw7W9/OztXSOe76Y1LV199dcybNy/+4R/+ocLwkUceiXSx2pPAL7F055pSpRt1pbii7Xfb/oCq\nOWerqIiCTugkrqAVn4qd7tz+zd/8TVnguOOOiwceeCC7C54C7TSkV1S1TfPcc89lVxT/+Z//Ofvj\nkO6gpz8s6UpkugOUgtH//u//jpNPPrm83SKOpNeitHVL4+mEOwWPqfOMNBx22GHxV3/1V9l4uiv5\nv//7v5FOdlKnMGn43Oc+F//yL/8Se++9dzbtn44FUkcv6eJQaRg0aFB8+MMfzu6old4vn07G3/KW\nt2SdHX7961+P1MOooXcE2NbuOHbs2Oxi6Hve855Id3c7Gqr5fe5ovaLPS53BpTu36ZVq6W5NR0Nq\nxfSjH/0o0on6SSedFOkVbK9+9as7SmpeG4HUOWS6WN/2b1sKItMrwP70pz+VU/rbVaaoamS//fbL\nLt7fcccd2d3xtFK6a5laJKa/b6XhO9/5Tnb+kFoypiEFmIceemh2U+QVr3hFKVnhv7v7DfjZz36W\n3UxK5wlpeNWrXpUdv+m1aykITUOqi/Rq4dRSwfCSQHeuKVXq0LDt70NHdtWcs3W0XtHmuYNetBrv\npry77LJLl3fA05Xe0nsNU4/Dqdl2Cs5Lw/Dhw7PAvTTt+yWBZJV6wn/hhRc6JWlrmxLdfffd8fOf\n/zze+MY3drqOBVsLpGM4ndy8+OKL2cJ0USkdp+nk8thjj81eY+WK+NZuPZnDtidqER/4wAeyC3Gp\niWotQ3e/z7Vsq1HT/tu//Vv8+te/zt4n31kZU1CZHutKj3FNnDjRXfTOoNrNTxfeU7Pr0pB+R9N0\nemyj7eBvV1uN7sfTheR0oShd/CgNKUBMgWPbt7wcffTRWVBeSpPOxdauXRvLli0rzfK9RaC734B0\n8fOJLS3p2jZp/9CHPhQzZ84s3y1PTdvTowVf/vKXmf6fQHeuPYVqf87W0+002nruoDdajfawPKlZ\nemr+m559/M///M/yVtIfiXQXNwU8Tz/9dNZ0pfQsX3p2p+2VsrT+/fffH+mOmuHPAkuWLMmaqqeA\nMbVSSJZpuOaaa2LMmDFZIJlaIqQrtaeddlp5xdQyIQ2auJdJuhxJJ4upBci0adMiPWJQuqOQjul0\nESkdq6l1QmoSmC58eB69S86stcdHP/rRikSplUdqVVMa2JYkavtOrTxS0+rXvva18d3vfjfe9a53\ndbmBzn6fu1ypoAvTqytTa66LLroo+x3o6A5Y+huVWomli6KpFcO//uu/xpe+9KUYPNgpUVeHTbpY\nnI7F++67L3su95e//GXWN03bIDKt729XV4odL0sXP9Id3a985StZgnQRKb1XPp1XlYbUl03b4bLL\nLouDDz44Uos9w58FuvsNSH2pHHPMMZnbW9/61uxcIF2sSxfxS0OplV2p76DS/CJ/d+eabFIMkI7L\ntkO6q97+Il5a3tk5W9t1izzuDnqBaz8FjukKbPqkzl/SD1UKINOnNCxfvjweeuih7LN48eLsau2C\nBQtKi8vf9957b9asPV1hS821ij6k58tLtqk5a7oKe/nll0e6k14aHnvssbJt6kAjXQBJz0MZahNI\nf2iTdbJNnbmk4zo9glEa0jP+Z5xxRnbynTo6TI8YpOcjDV0LpLs6KYBp+yk9MlBak21JovbvdKxe\ncMEFkU4WV69evdUGqvl93molMzKBFKCn393k235Iv7U//elPsz5X0rLUOVT6O6eD0/ZSW0+nCxip\nE87Snd4URLbvNHLrtcypRiD1Q5MeE1i0aFGWPBl3FNSUtpXOtVKT91mzZpVm+W4j0NVvQOqkM13M\nT4apVWPqWDIFkW4utQHsZLQr17RK+ruVbuy1/Tz44IMVW+vunK0icYEnXC4ucOWnjst+8YtfZALD\nhg3LmlKlk/G2Q/pj/IUvfKE8K6VPdx7T/NKdiXQ3Mv0hSXchOur5ubxygUbS3Zn0Q5aG1DlGam3Q\n/nnI1DKh9Ax6Speau6aT9XTyaKheILX4SJ09peAxNZVKj1mUhtSpTjpm0x/j2bNnZ7NXrlyZdWqY\nmme3fTyjtI7vlwRe97rXle/mlEzS8fnkk09mk2xLKj3/Tj0J33DDDVlTyvS2h7ZDNb/PbdMb/7NA\nar6emqumvinS4wRth9RiIXXGlTpALf0mp+900S79XTN0LZD+1qemv+mTnuNPHUGl5/0NL08gXfxI\nzzynRwbSa1XT//8UNHY0XHjhhdlbX9LftvHjx3eUpPDzuvoNSI8EpD6B0m9A+qTfg3RTJXWUnILH\ndL5m6FigK9e0RmqpmCy7Gro6Z+tqvaItE6AXrcbblDf9Qaj1xz01p0pNtNNV3hSgp6Zs6e5kaobd\ntnl2m90UcjT9+NdqmwKidLJuqE1gr7326tQ63WFIV8vb3h1Lz50l63SiXupsp7Y9Sp0E2L784yA9\nC5kedTnqqKOyT9uebnvy+/zyc9Q4W0j/x1OP16nZcNv3GKdWH+nRjXPPPbdc2HSROV0gTXfXnZyX\nWTocScdqCmrSheTU8q6zILLDlc3sUiBd/Ehvd0iPEaTm7e2H1CQ4HbfJPjV9T3/7DJ0LdPQbkM5f\n06NuqeVB6pMmDemRo+SaHn97+OGH/QZ0Tpot6ci1m1UqFnd1zlaRsOATmrgX/ADorvjp9Smp6XX6\npKA83UEbNWpU9l7vP/7xj9lzfOnZvXSnopQu3Zk0dC9Qatb+hz/8IbsDkZ6dbvtMf/dbkKI7gXQy\nnp4xTUF66ZN6a04Xk7xeqTu9rpez7dqn2qVveMMb4qyzzsruRFa7jnTVCaTWX6kJ6+9///tshfTa\npNTfR+rlvfR7kL7P3PIqpdSaRhPX7l3THbT0GFy6uNlRENn9FqToTCB1Apdetfr1LW8a6ah5e/qb\n9b3vfS973W1q7Vg652p7AaqzbRd1fvvfgNTSLvX5kS7IpQtz6dWBqVXdxRdfnD3Old5IYOheoL1r\naY3UaWHpuGz7nZwNtQkI0GvzKlzqK664Imv6njqBSa8FS1cXUxOs9MxvasaSXleT7lCk5aVP2ybx\nhQOrocApGE9m6TVq6URn0qRJkbwNvSOQnntKnRmlFh7th7/927+N1MFRer2SoXYBtrWbdbVGumum\no6euhHq2LD020LaTzXRRKf3O7r///hUbTEFnehbdRbsKlk4nUodmqQ8Vz593StSjBalFTbqrm1on\npv5S2g+pL5vUX0LqLK50vpW+77nnnvZJTf+fQPvfgDQ7PXuejuEUqKflyTud16ZOUNNvgaF7gY5c\n01rXXXddxbFZOk5vvvnm7jcqRYVA05YmM60Vc0wQIECAAAECBAgQIECggQWWLl2aBekp4DQQyJOA\nAD1PtSEvBAgQIECAAAECBAgQIFBYAU3cC1v1Ck6AAAECBAgQIECAAAECeRIQoOepNuSFAAECBAgQ\nIECAAAECBAorIEAvbNUrOAECBAgQIECAAAECBAjkSUCAnqfakBcCBAgQIECAAAECBAgQKKyAAL2w\nVa/gBAgQIECAAAECBAgQIJAnAQF6nmpDXggQIECAAAECBAgQIECgsAIC9MJWvYITIECAAAECBAgQ\nIECAQJ4EBOh5qg15IUCAAAECBAgQIECAAIHCCgjQC1v1Ck6AAAECBAgQIECAAAECeRIQoOepNuSF\nAAECBAgQIECAAAECBAorIEAvbNUrOAECBAgQIECAAAECBAjkSUCAnqfakBcCBAgQIECAAAECBAgQ\nKKyAAL2wVa/gBAgQIECAAAECBAgQIJAnAQF6nmpDXggQIECAAAECBAgQIECgsAIC9MJWvYITIECA\nAAECBAgQIECAQJ4EBOh5qg15IUCAAAECBAgQIECAAIHCCgjQC1v1Ck6AAAECBAgQIECAAAECeRIQ\noOepNuSFAAECBAgQIECAAAECBAorIEAvbNUrOAECBAgQIECAAAECBAjkSUCAnqfakBcCBAgQIECA\nAAECBAgQKKyAAL2wVa/gBAgQIECAAAECBAgQIJAnAQF6nmpDXggQIECAAAECBAgQIECgsAIC9MJW\nvYITIECAAAECBAgQIECAQJ4EBOh5qg15IUCAAAECBAgQIECAAIHCCgjQC1v1Ck6AAAECBAgQIECA\nAAECeRIQoOepNuSFAAECBAgQIECAAAECBAorIEAvbNUrOAECBAgQIECAAAECBAjkSUCAnqfakBcC\nBAgQIECAAAECBAgQKKyAAL2wVa/gBAgQIECAAAECBAgQIJAnAQF6nmpDXggQIECAAAECBAgQIECg\nsAIC9MJWvYITIECAAAECBAgQIECAQJ4EBOh5qg15IUCAAAECBAgQIECAAIHCCgjQC1v1Ck6AAAEC\nBAgQIECAAAECeRIQoOepNuSFAAECBAgQIECAAAECBAorIEAvbNUrOAECBAgQIECAAAECBAjkSUCA\nnqfakBcCBAgQIECAAAECBAgQKKyAAL2wVa/gBAgQIECAAAECBAgQIJAnAQF6nmpDXggQIECAAAEC\nBAgQIECgsAIC9MJWvYITIECAAAECBAgQIECAQJ4EBOh5qg15IUCAAAECBAgQIECAAIHCCgjQC1v1\nCk6AAAECBAgQIECAAAECeRIQoOepNuSFAAECBAgQIECAAAECBAorIEAvbNUrOAECBAgQIECAAAEC\nBAjkSUCAnqfakBcCBAgQIECAAAECBAgQKKyAAL2wVa/gBAgQIECAAAECBAgQIJAngf8PV7EaTGjS\nswcAAAAASUVORK5CYII=\n"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R -w 1000 -h 500 -u px -i  NoverS_df  # this sets the size of the plot...otherwise, it will go off the page\n",
    "\n",
    "require(ggplot2)\n",
    "library(ggplot2)\n",
    "\n",
    "NoverS_df$gene = gsub(\"neuraminidase\",\"NA\", NoverS_df$gene)\n",
    "NoverS_df$genef = factor(NoverS_df$gene, levels=c('PB2','PB1','PA','HA','NP','NA','M1','M2','NS1','NEP'))\n",
    "\n",
    "p2 <- ggplot(data=NoverS_df, aes(x=genef, y=mean, color=species, fill=species)) + \n",
    "    geom_col(position=\"dodge\")+   \n",
    "    #geom_errorbar(data=NoverS_df, aes(x=genef, ymin=mean - std, ymax=mean+std), position=\"dodge\")+\n",
    "    labs(x=\"gene\",y=\"NS/S\")+\n",
    "    scale_color_manual(values=c(\"#99bfaa\",\"#5c3d46\"),breaks = c(\"human synonymous\",\"human nonsynonymous\",\"duck synonymous\",\"duck nonsynonymous\"))+\n",
    "    scale_fill_manual(values=c(\"#99bfaa\",\"#5c3d46\"),breaks = c(\"human synonymous\",\"human nonsynonymous\",\"duck synonymous\",\"duck nonsynonymous\"))+\n",
    "    theme(plot.title = element_text(size=20, hjust=0.5))+\n",
    "    #scale_y_log10(limits=c(-0.001,10000), breaks=c(0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000))+\n",
    "    theme(strip.text = element_text(size=16))+\n",
    "    theme(panel.grid.major.y=element_line(colour=NA))+\n",
    "    theme(panel.grid.minor=element_line(colour=NA,size=NA))+    \n",
    "    theme(strip.background = element_rect(colour=NA, fill=NA))+\n",
    "    theme(axis.line.x=element_line(colour=\"black\"))+\n",
    "    theme(axis.line.y=element_line(colour=\"black\"))+\n",
    "    theme(axis.title.y=element_text(size=16, vjust=8))+\n",
    "    theme(axis.title.x=element_text(size=16, vjust=-15))+\n",
    "    theme(axis.text=element_text(size=16, colour=\"black\"))+\n",
    "    theme(axis.text.x=element_text(size=13))+\n",
    "    theme(legend.text=element_text(size=16))+\n",
    "    theme(legend.title=element_blank())+\n",
    "    theme(panel.margin=unit(1, \"lines\"))+\n",
    "    theme(plot.margin=unit(c(1,1,1,1),\"cm\"))+\n",
    "    theme(legend.key.size=unit(0.7, \"cm\"))+\n",
    "    #theme(legend.key=element_rect(fill=NA,size = 2))+\n",
    "    theme(panel.background=element_rect(fill=NA))+\n",
    "    theme(legend.key=element_rect(fill=NA))\n",
    "\n",
    "p2\n",
    "#ggsave(\"Fig-5-NS_S_SNPs_per_site-2019-02-11.pdf\", p, width = 13, height = 6, device=pdf, path=\"/Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/paper-and-figure-drafts/figures-2019-01-08\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PA</th>\n",
       "      <th>NEP</th>\n",
       "      <th>HA</th>\n",
       "      <th>NP</th>\n",
       "      <th>neuraminidase</th>\n",
       "      <th>M2</th>\n",
       "      <th>PB1</th>\n",
       "      <th>PB2</th>\n",
       "      <th>NS1</th>\n",
       "      <th>M1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.201002</td>\n",
       "      <td>3496.381688</td>\n",
       "      <td>3.340370</td>\n",
       "      <td>0.602750</td>\n",
       "      <td>0.283898</td>\n",
       "      <td>4389.694987</td>\n",
       "      <td>0.284555</td>\n",
       "      <td>0.147074</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>750.213720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.551321</td>\n",
       "      <td>3496.872957</td>\n",
       "      <td>0.557052</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.283452</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.295446</td>\n",
       "      <td>0.098943</td>\n",
       "      <td>1728.110599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.137884</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>0.301236</td>\n",
       "      <td>951.636286</td>\n",
       "      <td>4405.286344</td>\n",
       "      <td>0.142796</td>\n",
       "      <td>1706.998695</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.613586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000233</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2251.958144</td>\n",
       "      <td>0.303851</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>8810.936900</td>\n",
       "      <td>0.142796</td>\n",
       "      <td>1.499936</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.000059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.415577</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.556224</td>\n",
       "      <td>0.228054</td>\n",
       "      <td>0.282711</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2830.712202</td>\n",
       "      <td>0.449637</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.153246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.092357</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.185411</td>\n",
       "      <td>0.000348</td>\n",
       "      <td>0.283055</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.552607</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.304044</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.300400</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000466</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.558318</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         PA          NEP           HA        NP  neuraminidase           M2  \\\n",
       "0  2.201002  3496.381688     3.340370  0.602750       0.283898  4389.694987   \n",
       "1  0.551321  3496.872957     0.557052  0.000173       0.283452     0.000064   \n",
       "2  0.137884     1.000000     0.000370  0.301236     951.636286  4405.286344   \n",
       "3  0.000233     1.000000  2251.958144  0.303851       0.000149  8810.936900   \n",
       "4  0.415577     1.000000     0.556224  0.228054       0.282711     1.000000   \n",
       "5  0.092357     1.000000     0.185411  0.000348       0.283055     1.000000   \n",
       "6  0.552607     1.000000     0.000185  0.304044       0.000297     1.000000   \n",
       "7  0.000466     1.000000     0.558318  1.000000       1.000000     1.000000   \n",
       "\n",
       "           PB1          PB2       NS1           M1  \n",
       "0     0.284555     0.147074  0.000154   750.213720  \n",
       "1     0.000126     0.295446  0.098943  1728.110599  \n",
       "2     0.142796  1706.998695  0.000154     0.613586  \n",
       "3     0.142796     1.499936  0.000154     0.000059  \n",
       "4  2830.712202     0.449637  1.000000     0.153246  \n",
       "5     1.000000     0.000525  1.000000     1.000000  \n",
       "6     1.000000     0.300400  1.000000     1.000000  \n",
       "7          NaN     0.000263  1.000000     1.000000  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "humanList = []\n",
    "\n",
    "geneList = list(set(human['gene']))\n",
    "\n",
    "for gene in geneList:\n",
    "    genedf = human[human['gene'] == gene]['N_over_S']\n",
    "    genedf = genedf.reset_index(drop=True)\n",
    "    \n",
    "    humanList.append(genedf)\n",
    "    \n",
    "    human_compare = pd.concat(humanList, axis=1)\n",
    "\n",
    "human_compare.columns = geneList\n",
    "human_compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PA</th>\n",
       "      <th>neuraminidase</th>\n",
       "      <th>HA</th>\n",
       "      <th>NP</th>\n",
       "      <th>M2</th>\n",
       "      <th>PB1</th>\n",
       "      <th>PB2</th>\n",
       "      <th>NEP</th>\n",
       "      <th>NS1</th>\n",
       "      <th>M1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.138628</td>\n",
       "      <td>0.283468</td>\n",
       "      <td>750.562922</td>\n",
       "      <td>0.301394</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.269297</td>\n",
       "      <td>3842.459174</td>\n",
       "      <td>1732.764071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000464</td>\n",
       "      <td>0.000297</td>\n",
       "      <td>751.691305</td>\n",
       "      <td>872.248397</td>\n",
       "      <td>1.0</td>\n",
       "      <td>566.211492</td>\n",
       "      <td>0.000525</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1732.601790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000466</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000370</td>\n",
       "      <td>0.182399</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.280050</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         PA  neuraminidase          HA          NP   M2         PB1       PB2  \\\n",
       "0  0.138628       0.283468  750.562922    0.301394  1.0    0.000505  0.000131   \n",
       "1  0.000464       0.000297  751.691305  872.248397  1.0  566.211492  0.000525   \n",
       "2  0.000466       1.000000    0.000370    0.182399  1.0    0.000252  1.000000   \n",
       "3  1.000000       1.000000    0.280050    1.000000  1.0    1.000000  1.000000   \n",
       "4  1.000000       1.000000    1.000000    1.000000  1.0    1.000000  1.000000   \n",
       "\n",
       "        NEP          NS1           M1  \n",
       "0  0.269297  3842.459174  1732.764071  \n",
       "1  1.000000     1.000000  1732.601790  \n",
       "2  1.000000     1.000000     1.000000  \n",
       "3  1.000000     1.000000     1.000000  \n",
       "4  1.000000     1.000000     1.000000  "
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duckList = []\n",
    "\n",
    "geneList = list(set(duck['gene']))\n",
    "\n",
    "for gene in geneList:\n",
    "    genedf = duck[duck['gene'] == gene]['N_over_S']\n",
    "    genedf = genedf.reset_index(drop=True)\n",
    "    \n",
    "    duckList.append(genedf)\n",
    "    \n",
    "    duck_compare = pd.concat(duckList, axis=1)\n",
    "\n",
    "duck_compare.columns = geneList\n",
    "duck_compare"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run t-tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([0.69335823]), pvalue=array([0.51070142]))"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_PB1 = human_compare[[\"PB1\"]].dropna()\n",
    "d_PB1 = duck_compare[['PB1']].dropna()\n",
    "stats.ttest_ind(h_PB1, d_PB1, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([0.99898941]), pvalue=array([0.35107294]))"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_PB2 = human_compare[[\"PB2\"]].dropna()\n",
    "d_PB2 = duck_compare[['PB2']].dropna()\n",
    "stats.ttest_ind(h_PB2, d_PB2, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([0.18956069]), pvalue=array([0.85322851]))"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_PA = human_compare[[\"PA\"]].dropna()\n",
    "d_PA = duck_compare[['PA']].dropna()\n",
    "stats.ttest_ind(h_PA, d_PA, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([-0.05521979]), pvalue=array([0.95697105]))"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_HA = human_compare[[\"HA\"]].dropna()\n",
    "d_HA = duck_compare[['HA']].dropna()\n",
    "stats.ttest_ind(h_HA, d_HA, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([-1.00159626]), pvalue=array([0.37321615]))"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_NP = human_compare[[\"NP\"]].dropna()\n",
    "d_NP = duck_compare[['NP']].dropna()\n",
    "stats.ttest_ind(h_NP, d_NP, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([0.99703824]), pvalue=array([0.35195523]))"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_NA = human_compare[[\"neuraminidase\"]].dropna()\n",
    "d_NA = duck_compare[[\"neuraminidase\"]].dropna()\n",
    "stats.ttest_ind(h_NA, d_NA, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([-0.80025168]), pvalue=array([0.45294882]))"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_M1 = human_compare[[\"M1\"]].dropna()\n",
    "d_M1 = duck_compare[['M1']].dropna()\n",
    "stats.ttest_ind(h_M1, d_M1, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([1.86976628]), pvalue=array([0.10371342]))"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_M2 = human_compare[[\"M2\"]].dropna()\n",
    "d_M2 = duck_compare[['M2']].dropna()\n",
    "stats.ttest_ind(h_M2, d_M2, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([-1.00063459]), pvalue=array([0.3736286]))"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_NS1 = human_compare[[\"NS1\"]].dropna()\n",
    "d_NS1 = duck_compare[['NS1']].dropna()\n",
    "stats.ttest_ind(h_NS1, d_NS1, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=array([1.52778062]), pvalue=array([0.17040845]))"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_NEP = human_compare[[\"NEP\"]].dropna()\n",
    "d_NEP = duck_compare[['NEP']].dropna()\n",
    "stats.ttest_ind(h_NEP, d_NEP, axis=0, equal_var=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Single gene results: \n",
    "\n",
    "These tests tested the hypothesis that the ratio of NS/S is different in humans and ducks in each gene. No gene showed any evidence of a significant difference in the ratio of NS and S diversity between humans and ducks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "H5N1_v2",
   "language": "python",
   "name": "h5n1_v2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
