{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Develop a vcf annotater that will translate SNPs into amino acid changes\n",
    "\n",
    "May 1, 2018 \n",
    "\n",
    "I was having trouble with SNPEff mis-annotating coding region changes in M2 and NEP, so I decided to write my own. This is a simple notebook that will read in the reference sequences that reads were mapped to, a gtf file for those references sequences with coding region coordinates, and a vcf file output by Varscan. It will then determine whether SNVs occur within a coding region, and if so, annotate as amino acid changing or not. Currently, this will not annotate SNVs that fall outside of a coding region. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, subprocess, glob, os, shutil, re, importlib, csv\n",
    "from subprocess import call\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from Bio import AlignIO\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary convert single letter to 3 letter amino acid symbols\n",
    "amino_acid_abbreviations = {\"A\": \"Ala\",\"R\": \"Arg\",\"N\": \"Asn\",\"D\": \"Asp\",\"C\": \"Cys\",\"Q\": \"Gln\",\"E\": \"Glu\",\"G\": \"Gly\",\"H\": \"His\",\"I\": \"Ile\",\"L\": \"Leu\",\"K\": \"Lys\",\"M\": \"Met\",\"F\": \"Phe\",\"P\": \"Pro\",\"O\": \"Pyl\",\"S\": \"Ser\",\"U\": \"Sec\",\"T\": \"Thr\",\"W\": \"Trp\",\"Y\": \"Tyr\",\"V\": \"Val\", \"B\":\"Asx\",\"Z\":\"Glx\",\"U\":\"Sec\",\"X\":\"Xaa\",\"J\":\"Xle\",\"*\":\"Stop\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in reference sequences and gtfs and store information about protein sequences\n",
    "\n",
    "def define_coding_regions(gtf_file):\n",
    "    with open(gtf_file, \"r\") as gtf:\n",
    "        coding_regions = {}\n",
    "        \n",
    "        for line in gtf:\n",
    "            if line.strip(\"\\n\") != \"\":      # ignore blank lines (otherwise throws an index error)\n",
    "                sequence_name = line.split(\"\\t\")[0]\n",
    "                annotation_type = line.split(\"\\t\")[2]\n",
    "                start = int(line.split(\"\\t\")[3]) - 1  # adding the -1 here for 0 indexing\n",
    "                stop = int(line.split(\"\\t\")[4]) - 1    # adding the -1 here for 0 indexing\n",
    "                gene_name = line.split(\"\\t\")[8]\n",
    "                gene_name = gene_name.split(\";\")[0]\n",
    "                gene_name = gene_name.replace(\"gene_id \",\"\")\n",
    "                gene_name = gene_name.replace(\"\\\"\",\"\")\n",
    "            \n",
    "                if annotation_type.lower() == \"cds\":\n",
    "                    if sequence_name not in coding_regions:\n",
    "                        coding_regions[sequence_name] = {}\n",
    "                        coding_regions[sequence_name][gene_name] = [start, stop]\n",
    "                    elif sequence_name in coding_regions and gene_name not in coding_regions[sequence_name]:\n",
    "                        coding_regions[sequence_name][gene_name] = [start, stop]\n",
    "                    elif gene_name in coding_regions[sequence_name]:\n",
    "                        coding_regions[sequence_name][gene_name].extend([start, stop])\n",
    "        \n",
    "            # sort coding region coordinates so that they are always in the correct order\n",
    "            for sequence_name in coding_regions:\n",
    "                for gene in coding_regions[sequence_name]:\n",
    "                    coding_regions[sequence_name][gene] = sorted(coding_regions[sequence_name][gene])\n",
    "                    \n",
    "    return(coding_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull in fasta file and store gene segments in a dictionary\n",
    "def read_fastas(fasta_file):\n",
    "    reference_sequence = {}\n",
    "    \n",
    "    for seq in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        sequence_name = str(seq.id)\n",
    "        sequence = str(seq.seq).lower()\n",
    "        reference_sequence[sequence_name] = sequence\n",
    "    \n",
    "    return(reference_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# divide up sequences into their coding regions \n",
    "def slice_fastas(coding_regions, reference_sequence, fasta_file):\n",
    "    transcripts = {}\n",
    "    stop_codons = [\"taa\",\"tag\",\"tga\"]\n",
    "    refseqname = fasta_file.split(\"/\")[7]\n",
    "    \n",
    "    for c in coding_regions: \n",
    "        for gene in coding_regions[c]:\n",
    "            transcripts[gene] = \"\"\n",
    "            coordinates = coding_regions[c][gene]   # define the coding regions for each gene\n",
    "            \n",
    "            for i in range(0,int(len(coordinates)),2):   # go along the coordinates in chunks of 2 at a time\n",
    "                sequence_chunk = reference_sequence[c][coordinates[i]:coordinates[i+1]+1]\n",
    "                transcripts[gene] += sequence_chunk     # append each piece of the transcript together \n",
    "    \n",
    "    # loop through each transcript to make sure that it begins with a start codon and ends with a stop codon\n",
    "    for t in transcripts: \n",
    "        if transcripts[t][0:3] != \"atg\":\n",
    "            print(\"WARNING! \" + refseqname + \" \" + t + \"does not contain a start codon!\")\n",
    "        if transcripts[t][-3:] not in stop_codons:\n",
    "            print(\"WARNING! \" + refseqname + \" \" + t + \" does not contain a stop codon! These are the last 3 nucleotides: \" + transcripts[t][-3:])\n",
    "            \n",
    "    #print(transcripts)\n",
    "    return(transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open vcf, read in SNPs, and annotate with amino acid changes \n",
    "\n",
    "def annotate_amino_acid_changes(coding_regions, reference_sequence, transcripts, vcf, aa_abreviations):\n",
    "    \n",
    "    with open(vcf, \"r\") as csvfile:\n",
    "        outfilename = vcf.replace(\".vcf\",\".LHM-annotated-variants.txt\")\n",
    "        \n",
    "        with open(outfilename, \"w\") as outfile:\n",
    "            to_write = [\"sample\",\"gene\",\"reference_position\",\"reference_allele\",\"variant_allele\",\"coding_region_change\",\"synonymous/nonsynonymous\",\"frequency(%)\",\"frequency\",\"\\n\"]\n",
    "            to_write2 = \"\\t\".join(to_write)\n",
    "            outfile.write(to_write2)\n",
    "        \n",
    "        reader = csv.reader(csvfile, delimiter=\"\\t\")    \n",
    "        for row in reader:\n",
    "            \n",
    "            # ignore comment lines\n",
    "            if \"##\" not in row[0] and \"#CHROM\" not in row[0]:\n",
    "                \n",
    "                sequence_name = row[0]\n",
    "                site = int(row[1]) - 1     # to make this 0 indexed\n",
    "                reference_allele = row[3].lower()\n",
    "                alternative_allele = row[4].lower()\n",
    "                \n",
    "                # pull out the frequency using a string search\n",
    "                SearchStr = '.+\\:([0-9]{1,2}\\.{0,1}[0-9]{0,2}\\%)'\n",
    "                result = re.search(SearchStr,row[9])\n",
    "                if result:\n",
    "                    frequency = '\\t'.join(result.groups())\n",
    "                    frequency_decimal = frequency.replace(\"%\",\"\")\n",
    "                    frequency_decimal = (float(frequency_decimal))/100\n",
    "                    frequency_decimal = \"%.4f\" % frequency_decimal    # only include 4 numbers after decimal\n",
    "\n",
    "                else:\n",
    "                    frequency = \"none reported\"\n",
    "\n",
    "                # figure out whether the SNP lies within a coding region: \n",
    "                for gene in coding_regions[sequence_name]:\n",
    "                    coordinates = coding_regions[sequence_name][gene]\n",
    "\n",
    "                    # go through gene coordinates 2 at a time; this is for genes with multiple regions \n",
    "                    for i in range(0,int(len(coordinates)),2):\n",
    "                        if site >= coordinates[i] and site <= coordinates[i+1]:  # if site is within the gene\n",
    "                            \n",
    "                            # determine the coding region site, depending on if there are 2 frames or 1 \n",
    "                            if len(coordinates) == 2:\n",
    "                                cds_site = site - coordinates[i]\n",
    "                            elif len(coordinates) == 4:\n",
    "                                cds_site = (coordinates[1] - coordinates[0]) + (site - coordinates[2] + 1)\n",
    "                                #print(gene,coordinates,site,cds_site)\n",
    "                                \n",
    "                            # now determine whether the site is in the 1st, 2nd, or 3rd codon position\n",
    "                            # if SNP is in 1st position in codon:\n",
    "                            aa_site = int(cds_site/3)+1\n",
    "                            \n",
    "                            if float(cds_site) % 3 == 0:\n",
    "                                codon = transcripts[gene][cds_site:cds_site+3]\n",
    "                                variant_codon = alternative_allele + codon[1:3]\n",
    "                                variant_aa = Seq(variant_codon).translate()\n",
    "                                ref_codon = reference_allele + codon[1:3]\n",
    "                                ref_aa = Seq(ref_codon).translate()\n",
    "                            \n",
    "                            # if variant is in the middle of the codon:    \n",
    "                            elif float(cds_site - 1) % 3 == 0:\n",
    "                                codon = transcripts[gene][cds_site-1:cds_site+2]\n",
    "                                variant_codon = codon[0] + alternative_allele + codon[2]\n",
    "                                variant_aa = Seq(variant_codon).translate()\n",
    "                                ref_codon = codon[0] + reference_allele + codon[2]\n",
    "                                ref_aa = Seq(ref_codon).translate()\n",
    "                            \n",
    "                            # if the variant is in the 3rd codon position\n",
    "                            elif float(cds_site -2) % 3 == 0:\n",
    "                                codon = transcripts[gene][cds_site-2:cds_site+1]\n",
    "                                variant_codon = codon[0:2] + alternative_allele\n",
    "                                variant_aa = Seq(variant_codon).translate()\n",
    "                                ref_codon = codon[0:2] + reference_allele\n",
    "                                ref_aa = Seq(ref_codon).translate()\n",
    "                                \n",
    "                                \n",
    "                            # return the amino acid changes, with single letters converted to 3-letter aa codes\n",
    "                            if ref_aa == variant_aa:\n",
    "                                syn_nonsyn = \"synonymous\"\n",
    "                            elif ref_aa != variant_aa and variant_aa == \"*\":\n",
    "                                syn_nonsyn = \"stop_gained\"\n",
    "                            elif ref_aa != variant_aa:\n",
    "                                syn_nonsyn = \"nonsynonymous\"\n",
    "                            \n",
    "                            amino_acid_change = amino_acid_abbreviations[ref_aa] + str(aa_site) + amino_acid_abbreviations[variant_aa]\n",
    "                        \n",
    "                        #else: \n",
    "                            #amino_acid_change = \"outside coding region\"\n",
    "                            #syn_nonsyn = \"outside coding region\"\n",
    "                            \n",
    "                                                \n",
    "                         \n",
    "                            with open(outfilename, \"a\") as outfile: \n",
    "                                output = [sequence_name,gene,str(site+1),reference_allele.upper(),alternative_allele.upper(),amino_acid_change,syn_nonsyn,frequency,frequency_decimal,\"\\n\"]\n",
    "                                output2 = \"\\t\".join(output)\n",
    "                                outfile.write(output2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in vcfs, gtfs, and fasta files into a single dictionary based on sampleid\n",
    "\n",
    "vcfs_directory = \"/Volumes/gradschool-and-postdoc-backups/post-doc/stored_files_too_big_for_laptop/H5N1_Cambodia_outbreak_study/Cambodia_H5_sequence_raw_data/combined_human_and_bird_usable_subset/\"\n",
    "gtfs_directory = \"/Volumes/gradschool-and-postdoc-backups/post-doc/stored_files_too_big_for_laptop/H5N1_Cambodia_outbreak_study/Cambodia_H5_sequence_raw_data/combined_human_and_bird_usable_subset/combined_vcfs_nodups/gtfs/\"\n",
    "sampleids = {}\n",
    "\n",
    "for f in glob.glob(vcfs_directory + \"*/coverage_norm_and_duplicate_read_removal/*.varscan0.01.vcf\"):  #coverage_norm_and_duplicate_read_removal/\n",
    "    sampleid = f.replace(vcfs_directory, \"\")\n",
    "    sampleid = sampleid.split(\"/\")[2]\n",
    "    sampleid = \"_\".join(sampleid.split(\"_\")[1:-3])\n",
    "    sampleid = sampleid.replace(\"_ori2\",\"\")\n",
    "    sampleid = sampleid.replace(\"CAMBODIA\",\"Cambodia\")\n",
    "    sampleid = sampleid.replace(\"Combodia\",\"Cambodia\")\n",
    "    \n",
    "    sampleids[sampleid] = {}\n",
    "    sampleids[sampleid][\"vcf\"] = f\n",
    "\n",
    "for f in glob.glob(gtfs_directory + \"*\"):\n",
    "    sampleid = f.replace(gtfs_directory, \"\")\n",
    "    sampleid = \"_\".join(sampleid.split(\"_\")[1:])\n",
    "    sampleid = sampleid.replace(\"_ori2\",\"\")\n",
    "    \n",
    "    if sampleid != \"duck_Guangdong_173_2004\" and sampleid != \"I_didnt_use\":\n",
    "        for g in glob.glob(f + \"*/genes.gtf\"):\n",
    "            sampleids[sampleid][\"gtf\"] = g\n",
    "        for h in glob.glob(f + \"*/sequences.fa\"):\n",
    "            sampleids[sampleid][\"fasta\"] = h\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sampleids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmoncla/anaconda/lib/python3.6/site-packages/Bio/Seq.py:150: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING! combined_human_and_bird_usable_subset HA does not contain a stop codon! These are the last 3 nucleotides: tgc\n"
     ]
    }
   ],
   "source": [
    "# run it!\n",
    "for s in sampleids:\n",
    "    coding_regions = define_coding_regions(sampleids[s][\"gtf\"])\n",
    "    reference_sequence = read_fastas(sampleids[s][\"fasta\"])\n",
    "    transcripts = slice_fastas(coding_regions, reference_sequence, sampleids[s]['fasta'])\n",
    "    vcf = sampleids[s][\"vcf\"]\n",
    "    annotate_amino_acid_changes(coding_regions, reference_sequence, transcripts, vcf, amino_acid_abbreviations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I ran this from the terminal to combine everything\n",
    "\n",
    "# for f in */coverage_norm_and_duplicate_read_removal/*.LHM-annotated-variants.txt; do cp $f /Users/lmoncla/Documents/H5N1_Cambodian_outbreak_study/Cambodia_H5_sequence_raw_data/combined_human_and_bird_usable_subset/combined_vcfs_nodups; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine all vcfs in this folder together into 1 parseable file\n",
    "output_directory = \"/Volumes/gradschool-and-postdoc-backups/post-doc/stored_files_too_big_for_laptop/H5N1_Cambodia_outbreak_study/Cambodia_H5_sequence_raw_data/combined_human_and_bird_usable_subset/combined_vcfs_nodups/\"\n",
    "vcf_outs = []\n",
    "outfilename = output_directory + \"combined_variants_nodups_2019-06-03.txt\"\n",
    "\n",
    "for f in glob.glob(output_directory + \"*.txt\"):\n",
    "    vcf_outs.append(f)\n",
    "    \n",
    "with open(outfilename, \"w\") as outfile:\n",
    "    header = [\"sampleid\",\"sample\",\"gene\",\"reference_position\",\"reference_allele\",\"variant_allele\",\"coding_region_change\",\"synonymous_nonsynonymous\",\"frequency(%)\",\"frequency\",\"\\n\"]\n",
    "    header = \"\\t\".join(header)\n",
    "    outfile.write(header)\n",
    "    \n",
    "    for v in vcf_outs:\n",
    "        with open(v, \"r\") as infile:\n",
    "            for line in infile:\n",
    "                if \"sample\" not in line:\n",
    "                    sample = \"/\".join(line.split(\"\\t\")[0].split(\"_\")[1:-1])\n",
    "                    sampleid = line.split(\"\\t\")[0]\n",
    "                    rest = \"\\t\".join(line.split(\"\\t\")[1:])\n",
    "                    outfile.write(sampleid+\"\\t\"+sample+\"\\t\"+rest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "H5N1_v2",
   "language": "python",
   "name": "h5n1_v2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
